// @udemy/react-navigation-components/dist/esm/pill-group/pill-group.react-component.js
import { Button } from "@udemy/react-core-components";
import classNames from "classnames";
import React from "react";
import styles from "./pill-group.module.css";
/**
 * The PillGroup component.
 *
 * @remarks
 * This is essentially a stylized unordered list (`ul`).
 */

export const PillGroup = (_ref) => {
  let { children, ...props } = _ref;
  return /*#__PURE__*/ React.createElement(
    "ul",
    Object.assign({}, props, {
      className: classNames(
        "ud-unstyled-list",
        styles["pill-group"],
        props.className
      ),
    }),
    children
  );
};
/**
 * An individual `Pill` to use within the `PillGroup` component
 *
 * @remark
 * This is an `HTMLLIElement` with a {@link Button} component within it.
 * Its `udStyle` is set to `secondary`.
 */

export const Pill = /*#__PURE__*/ React.forwardRef((_ref2, ref) => {
  let { udStyle = "secondary", ...props } = _ref2;
  return /*#__PURE__*/ React.createElement(
    "li",
    {
      className: styles.pill,
      ref: ref,
    },
    /*#__PURE__*/ React.createElement(
      Button,
      Object.assign(
        {
          udStyle: udStyle,
        },
        props,
        {
          round: true,
        }
      )
    )
  );
});
Pill.displayName = "Pill";
PillGroup.Pill = Pill;
