// @udemy/smart-bar/dist/esm/events.js
import { ClientEvent } from "@udemy/event-tracking";
export let NoticeType = /*#__PURE__*/ (function (NoticeType) {
  NoticeType["WEB_BANNER"] = "web_banner";
  NoticeType["FEATURED_BANNER"] = "featured_banner";
  NoticeType["HOME_BANNER"] = "home_banner";
  NoticeType["MOBILE_BANNER"] = "mobile_banner";
  NoticeType["SMART_BAR"] = "smart_bar";
  NoticeType["UFB_SMART_BAR"] = "ufb_smart_bar";
  NoticeType["INSTRUCTOR_BAR"] = "instructor_bar";
  NoticeType["CART_SUCCESS_MESSAGE"] = "cart_success_message";
  NoticeType["PURCHASE_SUCCESS_MESSAGE"] = "purchase_success_message";
  NoticeType["FALLBACK_BANNER"] = "fallback_banner";
  NoticeType["WEB_CAROUSEL_SLIDE"] = "web_carousel_slide";
  return NoticeType;
})({});
export class BannerImpressionEvent extends ClientEvent {
  constructor(_ref) {
    let {
      noticeId = 0,
      noticeType,
      personalizedNoticeSetId = null,
      personalizedNoticeSetName = null,
      topMembershipTargetGroupId = null,
      slideNumber = null,
    } = _ref;
    super("BannerImpressionEvent");
    this.noticeId = void 0;
    this.noticeType = void 0;
    this.personalizedNoticeSetId = void 0;
    this.personalizedNoticeSetName = void 0;
    this.topMembershipTargetGroupId = void 0;
    this.slideNumber = void 0;
    this.noticeId = noticeId;
    this.noticeType = noticeType;
    this.personalizedNoticeSetId = personalizedNoticeSetId;
    this.personalizedNoticeSetName = personalizedNoticeSetName;
    this.topMembershipTargetGroupId = topMembershipTargetGroupId;
    this.slideNumber = slideNumber;
  }
}
export class BannerClickEvent extends ClientEvent {
  constructor(_ref2) {
    let {
      noticeId = 0,
      noticeType,
      personalizedNoticeSetId = null,
      personalizedNoticeSetName = null,
      topMembershipTargetGroupId = null,
      slideNumber = null,
      url = null,
      uiRegion = null,
    } = _ref2;
    super("BannerClickEvent");
    this.noticeId = void 0;
    this.noticeType = void 0;
    this.personalizedNoticeSetId = void 0;
    this.personalizedNoticeSetName = void 0;
    this.topMembershipTargetGroupId = void 0;
    this.slideNumber = void 0;
    this.url = void 0;
    this.uiRegion = void 0;
    this.noticeId = noticeId;
    this.noticeType = noticeType;
    this.personalizedNoticeSetId = personalizedNoticeSetId;
    this.personalizedNoticeSetName = personalizedNoticeSetName;
    this.topMembershipTargetGroupId = topMembershipTargetGroupId;
    this.slideNumber = slideNumber;
    this.url = url;
    this.uiRegion = uiRegion;
  }
}
export class SmartbarImpressionEvent extends ClientEvent {
  constructor(_ref3) {
    let {
      noticeId = 0,
      noticeType,
      personalizedNoticeSetId = null,
      personalizedNoticeSetName = null,
      topMembershipTargetGroupId = null,
    } = _ref3;
    super("SmartbarImpressionEvent");
    this.noticeId = void 0;
    this.noticeType = void 0;
    this.personalizedNoticeSetId = void 0;
    this.personalizedNoticeSetName = void 0;
    this.topMembershipTargetGroupId = void 0;
    this.noticeId = noticeId;
    this.noticeType = noticeType;
    this.personalizedNoticeSetId = personalizedNoticeSetId;
    this.personalizedNoticeSetName = personalizedNoticeSetName;
    this.topMembershipTargetGroupId = topMembershipTargetGroupId;
  }
}
export class SmartbarClickEvent extends ClientEvent {
  constructor(_ref4) {
    let {
      noticeId = 0,
      noticeType,
      url,
      location = "action_url",
      personalizedNoticeSetId = null,
      personalizedNoticeSetName = null,
      topMembershipTargetGroupId = null,
    } = _ref4;
    super("SmartbarClickEvent");
    this.noticeId = void 0;
    this.noticeType = void 0;
    this.personalizedNoticeSetId = void 0;
    this.personalizedNoticeSetName = void 0;
    this.topMembershipTargetGroupId = void 0;
    this.url = void 0;
    this.location = void 0;
    this.noticeId = noticeId;
    this.noticeType = noticeType;
    this.personalizedNoticeSetId = personalizedNoticeSetId;
    this.personalizedNoticeSetName = personalizedNoticeSetName;
    this.topMembershipTargetGroupId = topMembershipTargetGroupId;
    this.url = url;
    this.location = location;
  }
}
export class SmartbarHideEvent extends ClientEvent {
  constructor(_ref5) {
    let {
      noticeId = 0,
      noticeType,
      personalizedNoticeSetId = null,
      personalizedNoticeSetName = null,
      topMembershipTargetGroupId = null,
    } = _ref5;
    super("SmartbarHideEvent");
    this.noticeId = void 0;
    this.noticeType = void 0;
    this.personalizedNoticeSetId = void 0;
    this.personalizedNoticeSetName = void 0;
    this.topMembershipTargetGroupId = void 0;
    this.noticeId = noticeId;
    this.noticeType = noticeType;
    this.personalizedNoticeSetId = personalizedNoticeSetId;
    this.personalizedNoticeSetName = personalizedNoticeSetName;
    this.topMembershipTargetGroupId = topMembershipTargetGroupId;
  }
}
export class DealOptInEvent extends ClientEvent {
  constructor() {
    let noticeId =
      arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
    super("DealOptInEvent");
    this.noticeId = void 0;
    this.noticeId = noticeId;
  }
}
