// @udemy/browse-course/dist/esm/async-price-course-card/async-price-course-card.react-component.js
import PropTypes from "prop-types";
import React, { Component } from "react";
import { BrowseCourseCard } from "../browse-course-card/browse-course-card.react-component";
import { AsyncCourseStaticPriceText } from "../price-text/async-course-static-price-text.react-component";
export class AsyncPriceCourseCard extends Component {
  constructor() {
    super(...arguments);
    this.renderWithAsyncCourseStaticPrice = (
      StaticPriceTextComponent,
      priceTextProps
    ) => {
      // AsyncCourseStaticPriceText derives these props internally from `course`, so
      // pluck them lest they overwrite AsyncCourseStaticPriceText.
      const {
        /* eslint-disable @typescript-eslint/no-unused-vars */
        listPrice,
        discountPrice,
        listPriceString,
        discountPriceString,
        /* eslint-enable @typescript-eslint/no-unused-vars */
        ...additionalPriceTextProps
      } = priceTextProps;
      return /*#__PURE__*/ React.createElement(
        AsyncCourseStaticPriceText,
        Object.assign(
          {
            courses: [this.props.course],
          },
          additionalPriceTextProps
        )
      );
    };
  }
  render() {
    return /*#__PURE__*/ React.createElement(
      BrowseCourseCard,
      Object.assign(
        {
          renderPriceText: this.renderWithAsyncCourseStaticPrice,
        },
        this.props
      )
    );
  }
}
AsyncPriceCourseCard.propTypes = {
  course: PropTypes.object.isRequired,
};
