// @udemy/browse-course/dist/esm/course-price-events/course-card-price-events.js
import { pricePropsFrom } from "@udemy/react-card-components";
import { PURCHASE_PRICE_TYPES } from "@udemy/shopping";

/** A subset of Course API for tracking */

/**
 * Injects `trackingEventContext` into `pricePropsFrom` function from
 * `@udemy/react-card-components`.
 *
 * @remarks
 *
 * Augments the existing convenience function to pluck course card price props
 * from the `course` object. Adds a tracking event context for the rendered
 * price text component.
 *
 * @see {@link `@udemy/browse-course#StaticPriceText`} and
 * {@link `@udemy/browse-course#DynamicPriceText`}.
 */
export const pricePropsWithTrackingFrom = (_ref) => {
  let { course, priceTextProps } = _ref;
  return {
    ...pricePropsFrom({
      course,
      priceTextProps,
    }),
    trackingEventContext: {
      buyableId: course.id,
      priceType: PURCHASE_PRICE_TYPES.individual_buyable,
      buyableType: "course",
      // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
      buyableTrackingId: course.frontendTrackingId || course.tracking_id,
    },
    ...priceTextProps,
  };
};
