// @udemy/protobuf-objects/build/protos/udemy/dto/user/v1beta/user_pb.js
// source: udemy/dto/user/v1beta/user.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require("google-protobuf");
var goog = jspb;
var global = Function("return this")();

var google_protobuf_timestamp_pb = require("google-protobuf/google/protobuf/timestamp_pb.js");
goog.object.extend(proto, google_protobuf_timestamp_pb);
goog.exportSymbol("proto.udemy.dto.user.v1.User", null, global);
goog.exportSymbol("proto.udemy.dto.user.v1.UserImages", null, global);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.udemy.dto.user.v1.UserImages = function (opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.udemy.dto.user.v1.UserImages, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.udemy.dto.user.v1.UserImages.displayName =
    "proto.udemy.dto.user.v1.UserImages";
}
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.udemy.dto.user.v1.User = function (opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.udemy.dto.user.v1.User, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.udemy.dto.user.v1.User.displayName = "proto.udemy.dto.user.v1.User";
}

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * Optional fields that are not set will be set to undefined.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
   * @param {boolean=} opt_includeInstance Deprecated. whether to include the
   *     JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.udemy.dto.user.v1.UserImages.prototype.toObject = function (
    opt_includeInstance
  ) {
    return proto.udemy.dto.user.v1.UserImages.toObject(
      opt_includeInstance,
      this
    );
  };

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Deprecated. Whether to include
   *     the JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.udemy.dto.user.v1.UserImages} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.udemy.dto.user.v1.UserImages.toObject = function (
    includeInstance,
    msg
  ) {
    var f,
      obj = {
        image50x50: jspb.Message.getFieldWithDefault(msg, 1, ""),
        image75x75: jspb.Message.getFieldWithDefault(msg, 2, ""),
        image100x100: jspb.Message.getFieldWithDefault(msg, 3, ""),
        image125H: jspb.Message.getFieldWithDefault(msg, 4, ""),
        image200H: jspb.Message.getFieldWithDefault(msg, 5, ""),
      };

    if (includeInstance) {
      obj.$jspbMessageInstance = msg;
    }
    return obj;
  };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.udemy.dto.user.v1.UserImages}
 */
proto.udemy.dto.user.v1.UserImages.deserializeBinary = function (bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.udemy.dto.user.v1.UserImages();
  return proto.udemy.dto.user.v1.UserImages.deserializeBinaryFromReader(
    msg,
    reader
  );
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.udemy.dto.user.v1.UserImages} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.udemy.dto.user.v1.UserImages}
 */
proto.udemy.dto.user.v1.UserImages.deserializeBinaryFromReader = function (
  msg,
  reader
) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
      case 1:
        var value = /** @type {string} */ (reader.readString());
        msg.setImage50x50(value);
        break;
      case 2:
        var value = /** @type {string} */ (reader.readString());
        msg.setImage75x75(value);
        break;
      case 3:
        var value = /** @type {string} */ (reader.readString());
        msg.setImage100x100(value);
        break;
      case 4:
        var value = /** @type {string} */ (reader.readString());
        msg.setImage125H(value);
        break;
      case 5:
        var value = /** @type {string} */ (reader.readString());
        msg.setImage200H(value);
        break;
      default:
        reader.skipField();
        break;
    }
  }
  return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.udemy.dto.user.v1.UserImages.prototype.serializeBinary = function () {
  var writer = new jspb.BinaryWriter();
  proto.udemy.dto.user.v1.UserImages.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.udemy.dto.user.v1.UserImages} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.udemy.dto.user.v1.UserImages.serializeBinaryToWriter = function (
  message,
  writer
) {
  var f = undefined;
  f = message.getImage50x50();
  if (f.length > 0) {
    writer.writeString(1, f);
  }
  f = message.getImage75x75();
  if (f.length > 0) {
    writer.writeString(2, f);
  }
  f = message.getImage100x100();
  if (f.length > 0) {
    writer.writeString(3, f);
  }
  f = message.getImage125H();
  if (f.length > 0) {
    writer.writeString(4, f);
  }
  f = message.getImage200H();
  if (f.length > 0) {
    writer.writeString(5, f);
  }
};

/**
 * optional string image_50x50 = 1;
 * @return {string}
 */
proto.udemy.dto.user.v1.UserImages.prototype.getImage50x50 = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.UserImages} returns this
 */
proto.udemy.dto.user.v1.UserImages.prototype.setImage50x50 = function (value) {
  return jspb.Message.setProto3StringField(this, 1, value);
};

/**
 * optional string image_75x75 = 2;
 * @return {string}
 */
proto.udemy.dto.user.v1.UserImages.prototype.getImage75x75 = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.UserImages} returns this
 */
proto.udemy.dto.user.v1.UserImages.prototype.setImage75x75 = function (value) {
  return jspb.Message.setProto3StringField(this, 2, value);
};

/**
 * optional string image_100x100 = 3;
 * @return {string}
 */
proto.udemy.dto.user.v1.UserImages.prototype.getImage100x100 = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.UserImages} returns this
 */
proto.udemy.dto.user.v1.UserImages.prototype.setImage100x100 = function (
  value
) {
  return jspb.Message.setProto3StringField(this, 3, value);
};

/**
 * optional string image_125_h = 4;
 * @return {string}
 */
proto.udemy.dto.user.v1.UserImages.prototype.getImage125H = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 4, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.UserImages} returns this
 */
proto.udemy.dto.user.v1.UserImages.prototype.setImage125H = function (value) {
  return jspb.Message.setProto3StringField(this, 4, value);
};

/**
 * optional string image_200_h = 5;
 * @return {string}
 */
proto.udemy.dto.user.v1.UserImages.prototype.getImage200H = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 5, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.UserImages} returns this
 */
proto.udemy.dto.user.v1.UserImages.prototype.setImage200H = function (value) {
  return jspb.Message.setProto3StringField(this, 5, value);
};

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * Optional fields that are not set will be set to undefined.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
   * @param {boolean=} opt_includeInstance Deprecated. whether to include the
   *     JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.udemy.dto.user.v1.User.prototype.toObject = function (
    opt_includeInstance
  ) {
    return proto.udemy.dto.user.v1.User.toObject(opt_includeInstance, this);
  };

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Deprecated. Whether to include
   *     the JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.udemy.dto.user.v1.User} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.udemy.dto.user.v1.User.toObject = function (includeInstance, msg) {
    var f,
      obj = {
        instant:
          (f = msg.getInstant()) &&
          google_protobuf_timestamp_pb.Timestamp.toObject(includeInstance, f),
        userId: jspb.Message.getFieldWithDefault(msg, 2, 0),
        title: jspb.Message.getFieldWithDefault(msg, 3, ""),
        name: jspb.Message.getFieldWithDefault(msg, 4, ""),
        displayName: jspb.Message.getFieldWithDefault(msg, 5, ""),
        email: jspb.Message.getFieldWithDefault(msg, 6, ""),
        status: jspb.Message.getFieldWithDefault(msg, 7, ""),
        image: jspb.Message.getFieldWithDefault(msg, 8, ""),
        timeZone: jspb.Message.getFieldWithDefault(msg, 9, ""),
        organizationId: jspb.Message.getFieldWithDefault(msg, 10, 0),
        country: jspb.Message.getFieldWithDefault(msg, 11, ""),
        locale: jspb.Message.getFieldWithDefault(msg, 12, ""),
        platform: jspb.Message.getFieldWithDefault(msg, 13, ""),
        isDisabled: jspb.Message.getBooleanFieldWithDefault(msg, 14, false),
        hasActivated: jspb.Message.getBooleanFieldWithDefault(msg, 15, false),
        isGenerated: jspb.Message.getBooleanFieldWithDefault(msg, 16, false),
        url: jspb.Message.getFieldWithDefault(msg, 17, ""),
        imageVersion: jspb.Message.getFieldWithDefault(msg, 18, 0),
        images:
          (f = msg.getImages()) &&
          proto.udemy.dto.user.v1.UserImages.toObject(includeInstance, f),
      };

    if (includeInstance) {
      obj.$jspbMessageInstance = msg;
    }
    return obj;
  };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.udemy.dto.user.v1.User}
 */
proto.udemy.dto.user.v1.User.deserializeBinary = function (bytes) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.udemy.dto.user.v1.User();
  return proto.udemy.dto.user.v1.User.deserializeBinaryFromReader(msg, reader);
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.udemy.dto.user.v1.User} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.udemy.dto.user.v1.User}
 */
proto.udemy.dto.user.v1.User.deserializeBinaryFromReader = function (
  msg,
  reader
) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
      case 1:
        var value = new google_protobuf_timestamp_pb.Timestamp();
        reader.readMessage(
          value,
          google_protobuf_timestamp_pb.Timestamp.deserializeBinaryFromReader
        );
        msg.setInstant(value);
        break;
      case 2:
        var value = /** @type {number} */ (reader.readUint64());
        msg.setUserId(value);
        break;
      case 3:
        var value = /** @type {string} */ (reader.readString());
        msg.setTitle(value);
        break;
      case 4:
        var value = /** @type {string} */ (reader.readString());
        msg.setName(value);
        break;
      case 5:
        var value = /** @type {string} */ (reader.readString());
        msg.setDisplayName(value);
        break;
      case 6:
        var value = /** @type {string} */ (reader.readString());
        msg.setEmail(value);
        break;
      case 7:
        var value = /** @type {string} */ (reader.readString());
        msg.setStatus(value);
        break;
      case 8:
        var value = /** @type {string} */ (reader.readString());
        msg.setImage(value);
        break;
      case 9:
        var value = /** @type {string} */ (reader.readString());
        msg.setTimeZone(value);
        break;
      case 10:
        var value = /** @type {number} */ (reader.readUint64());
        msg.setOrganizationId(value);
        break;
      case 11:
        var value = /** @type {string} */ (reader.readString());
        msg.setCountry(value);
        break;
      case 12:
        var value = /** @type {string} */ (reader.readString());
        msg.setLocale(value);
        break;
      case 13:
        var value = /** @type {string} */ (reader.readString());
        msg.setPlatform(value);
        break;
      case 14:
        var value = /** @type {boolean} */ (reader.readBool());
        msg.setIsDisabled(value);
        break;
      case 15:
        var value = /** @type {boolean} */ (reader.readBool());
        msg.setHasActivated(value);
        break;
      case 16:
        var value = /** @type {boolean} */ (reader.readBool());
        msg.setIsGenerated(value);
        break;
      case 17:
        var value = /** @type {string} */ (reader.readString());
        msg.setUrl(value);
        break;
      case 18:
        var value = /** @type {number} */ (reader.readUint64());
        msg.setImageVersion(value);
        break;
      case 19:
        var value = new proto.udemy.dto.user.v1.UserImages();
        reader.readMessage(
          value,
          proto.udemy.dto.user.v1.UserImages.deserializeBinaryFromReader
        );
        msg.setImages(value);
        break;
      default:
        reader.skipField();
        break;
    }
  }
  return msg;
};

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.udemy.dto.user.v1.User.prototype.serializeBinary = function () {
  var writer = new jspb.BinaryWriter();
  proto.udemy.dto.user.v1.User.serializeBinaryToWriter(this, writer);
  return writer.getResultBuffer();
};

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.udemy.dto.user.v1.User} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.udemy.dto.user.v1.User.serializeBinaryToWriter = function (
  message,
  writer
) {
  var f = undefined;
  f = message.getInstant();
  if (f != null) {
    writer.writeMessage(
      1,
      f,
      google_protobuf_timestamp_pb.Timestamp.serializeBinaryToWriter
    );
  }
  f = message.getUserId();
  if (f !== 0) {
    writer.writeUint64(2, f);
  }
  f = message.getTitle();
  if (f.length > 0) {
    writer.writeString(3, f);
  }
  f = message.getName();
  if (f.length > 0) {
    writer.writeString(4, f);
  }
  f = message.getDisplayName();
  if (f.length > 0) {
    writer.writeString(5, f);
  }
  f = message.getEmail();
  if (f.length > 0) {
    writer.writeString(6, f);
  }
  f = message.getStatus();
  if (f.length > 0) {
    writer.writeString(7, f);
  }
  f = message.getImage();
  if (f.length > 0) {
    writer.writeString(8, f);
  }
  f = message.getTimeZone();
  if (f.length > 0) {
    writer.writeString(9, f);
  }
  f = message.getOrganizationId();
  if (f !== 0) {
    writer.writeUint64(10, f);
  }
  f = message.getCountry();
  if (f.length > 0) {
    writer.writeString(11, f);
  }
  f = message.getLocale();
  if (f.length > 0) {
    writer.writeString(12, f);
  }
  f = message.getPlatform();
  if (f.length > 0) {
    writer.writeString(13, f);
  }
  f = message.getIsDisabled();
  if (f) {
    writer.writeBool(14, f);
  }
  f = message.getHasActivated();
  if (f) {
    writer.writeBool(15, f);
  }
  f = message.getIsGenerated();
  if (f) {
    writer.writeBool(16, f);
  }
  f = message.getUrl();
  if (f.length > 0) {
    writer.writeString(17, f);
  }
  f = message.getImageVersion();
  if (f !== 0) {
    writer.writeUint64(18, f);
  }
  f = message.getImages();
  if (f != null) {
    writer.writeMessage(
      19,
      f,
      proto.udemy.dto.user.v1.UserImages.serializeBinaryToWriter
    );
  }
};

/**
 * optional google.protobuf.Timestamp instant = 1;
 * @return {?proto.google.protobuf.Timestamp}
 */
proto.udemy.dto.user.v1.User.prototype.getInstant = function () {
  return /** @type{?proto.google.protobuf.Timestamp} */ (
    jspb.Message.getWrapperField(
      this,
      google_protobuf_timestamp_pb.Timestamp,
      1
    )
  );
};

/**
 * @param {?proto.google.protobuf.Timestamp|undefined} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setInstant = function (value) {
  return jspb.Message.setWrapperField(this, 1, value);
};

/**
 * Clears the message field making it undefined.
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.clearInstant = function () {
  return this.setInstant(undefined);
};

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.udemy.dto.user.v1.User.prototype.hasInstant = function () {
  return jspb.Message.getField(this, 1) != null;
};

/**
 * optional uint64 user_id = 2;
 * @return {number}
 */
proto.udemy.dto.user.v1.User.prototype.getUserId = function () {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 2, 0));
};

/**
 * @param {number} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setUserId = function (value) {
  return jspb.Message.setProto3IntField(this, 2, value);
};

/**
 * optional string title = 3;
 * @return {string}
 */
proto.udemy.dto.user.v1.User.prototype.getTitle = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 3, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setTitle = function (value) {
  return jspb.Message.setProto3StringField(this, 3, value);
};

/**
 * optional string name = 4;
 * @return {string}
 */
proto.udemy.dto.user.v1.User.prototype.getName = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 4, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setName = function (value) {
  return jspb.Message.setProto3StringField(this, 4, value);
};

/**
 * optional string display_name = 5;
 * @return {string}
 */
proto.udemy.dto.user.v1.User.prototype.getDisplayName = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 5, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setDisplayName = function (value) {
  return jspb.Message.setProto3StringField(this, 5, value);
};

/**
 * optional string email = 6;
 * @return {string}
 */
proto.udemy.dto.user.v1.User.prototype.getEmail = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 6, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setEmail = function (value) {
  return jspb.Message.setProto3StringField(this, 6, value);
};

/**
 * optional string status = 7;
 * @return {string}
 */
proto.udemy.dto.user.v1.User.prototype.getStatus = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 7, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setStatus = function (value) {
  return jspb.Message.setProto3StringField(this, 7, value);
};

/**
 * optional string image = 8;
 * @return {string}
 */
proto.udemy.dto.user.v1.User.prototype.getImage = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 8, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setImage = function (value) {
  return jspb.Message.setProto3StringField(this, 8, value);
};

/**
 * optional string time_zone = 9;
 * @return {string}
 */
proto.udemy.dto.user.v1.User.prototype.getTimeZone = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 9, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setTimeZone = function (value) {
  return jspb.Message.setProto3StringField(this, 9, value);
};

/**
 * optional uint64 organization_id = 10;
 * @return {number}
 */
proto.udemy.dto.user.v1.User.prototype.getOrganizationId = function () {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 10, 0));
};

/**
 * @param {number} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setOrganizationId = function (value) {
  return jspb.Message.setProto3IntField(this, 10, value);
};

/**
 * optional string country = 11;
 * @return {string}
 */
proto.udemy.dto.user.v1.User.prototype.getCountry = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 11, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setCountry = function (value) {
  return jspb.Message.setProto3StringField(this, 11, value);
};

/**
 * optional string locale = 12;
 * @return {string}
 */
proto.udemy.dto.user.v1.User.prototype.getLocale = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 12, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setLocale = function (value) {
  return jspb.Message.setProto3StringField(this, 12, value);
};

/**
 * optional string platform = 13;
 * @return {string}
 */
proto.udemy.dto.user.v1.User.prototype.getPlatform = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 13, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setPlatform = function (value) {
  return jspb.Message.setProto3StringField(this, 13, value);
};

/**
 * optional bool is_disabled = 14;
 * @return {boolean}
 */
proto.udemy.dto.user.v1.User.prototype.getIsDisabled = function () {
  return /** @type {boolean} */ (
    jspb.Message.getBooleanFieldWithDefault(this, 14, false)
  );
};

/**
 * @param {boolean} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setIsDisabled = function (value) {
  return jspb.Message.setProto3BooleanField(this, 14, value);
};

/**
 * optional bool has_activated = 15;
 * @return {boolean}
 */
proto.udemy.dto.user.v1.User.prototype.getHasActivated = function () {
  return /** @type {boolean} */ (
    jspb.Message.getBooleanFieldWithDefault(this, 15, false)
  );
};

/**
 * @param {boolean} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setHasActivated = function (value) {
  return jspb.Message.setProto3BooleanField(this, 15, value);
};

/**
 * optional bool is_generated = 16;
 * @return {boolean}
 */
proto.udemy.dto.user.v1.User.prototype.getIsGenerated = function () {
  return /** @type {boolean} */ (
    jspb.Message.getBooleanFieldWithDefault(this, 16, false)
  );
};

/**
 * @param {boolean} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setIsGenerated = function (value) {
  return jspb.Message.setProto3BooleanField(this, 16, value);
};

/**
 * optional string url = 17;
 * @return {string}
 */
proto.udemy.dto.user.v1.User.prototype.getUrl = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 17, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setUrl = function (value) {
  return jspb.Message.setProto3StringField(this, 17, value);
};

/**
 * optional uint64 image_version = 18;
 * @return {number}
 */
proto.udemy.dto.user.v1.User.prototype.getImageVersion = function () {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 18, 0));
};

/**
 * @param {number} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setImageVersion = function (value) {
  return jspb.Message.setProto3IntField(this, 18, value);
};

/**
 * optional UserImages images = 19;
 * @return {?proto.udemy.dto.user.v1.UserImages}
 */
proto.udemy.dto.user.v1.User.prototype.getImages = function () {
  return /** @type{?proto.udemy.dto.user.v1.UserImages} */ (
    jspb.Message.getWrapperField(this, proto.udemy.dto.user.v1.UserImages, 19)
  );
};

/**
 * @param {?proto.udemy.dto.user.v1.UserImages|undefined} value
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.setImages = function (value) {
  return jspb.Message.setWrapperField(this, 19, value);
};

/**
 * Clears the message field making it undefined.
 * @return {!proto.udemy.dto.user.v1.User} returns this
 */
proto.udemy.dto.user.v1.User.prototype.clearImages = function () {
  return this.setImages(undefined);
};

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.udemy.dto.user.v1.User.prototype.hasImages = function () {
  return jspb.Message.getField(this, 19) != null;
};

goog.object.extend(exports, proto.udemy.dto.user.v1);
