// @udemy/organization/dist/esm/insights/events.js
import { ClientEvent } from "@udemy/event-tracking";

/* --- Events from some page TO Team Hub / My Groups page --- */

// Event fired when an UFB Admin or Group Admin clicks on a link to the Team Hub page from the
// dedicated Homepage dashboard, specifically on a link to view insights for a group.
export class UFBTeamHubHomePageViewInsightsLinkClickEvent extends ClientEvent {
  constructor(_ref) {
    let { groupId } = _ref;
    super("UFBTeamHubHomePageViewInsightsLinkClickEvent");
    this.groupId = void 0;
    this.groupId = groupId;
  }
}
// Event fired when an UFB Admin or Group Admin clicks on a link to the Team Hub page from the
// dedicated Homepage dashboard, specifically on a link to view top topics for a group.
export class UFBTeamHubHomePageViewTopicsLinkClickEvent extends ClientEvent {
  constructor(_ref2) {
    let { groupId } = _ref2;
    super("UFBTeamHubHomePageViewTopicsLinkClickEvent");
    this.groupId = void 0;
    this.groupId = groupId;
  }
}
// Event fired when an UFB Admin or Group Admin clicks on a link to the Team Hub page from the
// dedicated Homepage dashboard, specifically on a link to discover more about Team Hub.
export class UFBTeamHubHomePageDiscoverMoreLinkClickEvent extends ClientEvent {
  constructor(_ref3) {
    let { groupId } = _ref3;
    super("UFBTeamHubHomePageDiscoverMoreLinkClickEvent");
    this.groupId = void 0;
    this.groupId = groupId;
  }
}
// Event fired when an UFB Admin or Group Admin clicks on a link to the Team Hub page from the
// Manage Menu.
export class UFBTeamHubManageMenuLinkClickEvent extends ClientEvent {
  constructor() {
    super("UFBTeamHubManageMenuLinkClickEvent");
  }
}
// Event fired when an UFB Admin or Group Admin clicks on a link to the Team Hub page from the
// Group Insights dashboard, specifically on the link related to a group name.
export class UFBTeamHubGroupInsightsGroupNameLinkClickEvent extends ClientEvent {
  constructor(_ref4) {
    let { groupId } = _ref4;
    super("UFBTeamHubGroupInsightsGroupNameLinkClickEvent");
    this.groupId = void 0;
    this.groupId = groupId;
  }
}
// Event fired when an UFB Admin or Group Admin clicks on a link to the Team Hub page from the
// Group Insights dashboard, specifically on a link to view activity for a group.
export class UFBTeamHubGroupInsightsViewActivityLinkClickEvent extends ClientEvent {
  constructor(_ref5) {
    let { groupId } = _ref5;
    super("UFBTeamHubGroupInsightsViewActivityLinkClickEvent");
    this.groupId = void 0;
    this.groupId = groupId;
  }
}
// Event fired when an UFB Admin or Group Admin clicks on a link to the Team Hub page from the
// Group Insights dashboard, specifically on a link to "MyGroups" in the header/banner.
export class UFBTeamHubGroupInsightsMyGroupsHeaderLinkClickEvent extends ClientEvent {
  constructor(_ref6) {
    let { groupId } = _ref6;
    super("UFBTeamHubGroupInsightsMyGroupsHeaderLinkClickEvent");
    this.groupId = void 0;
    this.groupId = groupId;
  }
}
// Event fired when an UFB Admin or Group Admin clicks on a link to the Team Hub page from the
// Group Insights dashboard, specifically on a link to "MyGroups" in the footer.
export class UFBTeamHubGroupInsightsMyGroupsFooterLinkClickEvent extends ClientEvent {
  constructor(_ref7) {
    let { groupId } = _ref7;
    super("UFBTeamHubGroupInsightsMyGroupsFooterLinkClickEvent");
    this.groupId = void 0;
    this.groupId = groupId;
  }
}

/* --- Events FROM Team Hub / My Groups page to some Insights page --- */

// Event fired when an UFB Admin or Group Admin clicks on a link to the User Activity page from
// the Team Hub page, specifically on a link in the Group learning overview widget to view more
// activity trends.
export class UFBTeamHubMyGroupsOverviewWidgetViewActivityTrendsLinkClickEvent extends ClientEvent {
  constructor(_ref8) {
    let { groupId } = _ref8;
    super("UFBTeamHubMyGroupsOverviewWidgetViewActivityTrendsLinkClickEvent");
    this.groupId = void 0;
    this.groupId = groupId;
  }
}
// Event fired when an UFB Admin or Group Admin clicks on a link to the Courses page from
// the Team Hub page, specifically on a link in the Top course widget to view active courses.
export class UFBTeamHubMyGroupsTopCourseWidgetActiveCoursesLinkClickEvent extends ClientEvent {
  constructor(_ref9) {
    let { groupId } = _ref9;
    super("UFBTeamHubMyGroupsTopCourseWidgetActiveCoursesLinkClickEvent");
    this.groupId = void 0;
    this.groupId = groupId;
  }
}
// Event fired when an UFB Admin or Group Admin clicks on a link to the User Activity page from
// the Team Hub page, specifically on a link in the Top course widget to view active users.
export class UFBTeamHubMyGroupsTopCourseWidgetActiveUsersLinkClickEvent extends ClientEvent {
  constructor(_ref10) {
    let { groupId } = _ref10;
    super("UFBTeamHubMyGroupsTopCourseWidgetActiveUsersLinkClickEvent");
    this.groupId = void 0;
    this.groupId = groupId;
  }
}
