// @udemy/react-header/dist/esm/mobile/mobile-nav/level-one/mobile-popular-topic-navs.react-component.js
import { observer } from "mobx-react";
import React from "react";
import { Tracker } from "@udemy/event-tracking";
import { CategoryNavItemSelectEvent } from "../../../events";
import { useMobileNavStore } from "../../../hooks/use-mobile-nav-store";
import {
  MobileNavItem,
  MobileNavL1Nav,
  MobileNavSection,
} from "../mobile-nav.react-component";
export const MobilePopularTopicNavs = observer(() => {
  var _mobileNavStore$mostP;
  const mobileNavStore = useMobileNavStore();
  function handleTopicNavClick(subcat, item) {
    const context = {
      categoryId: subcat.parentId,
      subcategoryId: subcat.id,
      topicId: item.id,
    };
    Tracker.publishEvent(
      new CategoryNavItemSelectEvent({
        context,
      })
    );
  }
  return /*#__PURE__*/ React.createElement(
    "span",
    null,
    (_mobileNavStore$mostP = mobileNavStore.mostPopularSubcategories) ===
      null || _mobileNavStore$mostP === void 0
      ? void 0
      : _mobileNavStore$mostP.map((subcat) => {
          var _subcat$popularTopics;
          return /*#__PURE__*/ React.createElement(
            MobileNavL1Nav,
            {
              key: `l1-subcat-id-${subcat.id}`,
              id: `header-toggle-side-nav-popular-topics-of-${subcat.id}`,
            },
            /*#__PURE__*/ React.createElement(
              MobileNavSection,
              null,
              (_subcat$popularTopics = subcat.popularTopics) === null ||
                _subcat$popularTopics === void 0
                ? void 0
                : _subcat$popularTopics.map((topic) =>
                    /*#__PURE__*/ React.createElement(
                      MobileNavItem,
                      {
                        onClick: () => handleTopicNavClick(subcat, topic),
                        key: `mobile-nav-topic-id-${topic.id}`,
                        "data-purpose": "topic-item",
                        href: topic.absolute_url || topic.url,
                      },
                      topic.title
                    )
                  )
            )
          );
        })
  );
});
