// @udemy/react-header/dist/esm/mobile/mobile-nav/level-one/mobile-instructor-performance-nav.react-component.js
import { observer } from "mobx-react";
import React from "react";
import { useI18n } from "@udemy/i18n";
import { useUDData } from "@udemy/ud-data";
import { useHeaderStore } from "../../../hooks/use-header-store";
import {
  MobileNavItem,
  MobileNavL1Nav,
  MobileNavSection,
} from "../mobile-nav.react-component";
export const MobileInstructorPerformanceNav = observer(() => {
  const headerStore = useHeaderStore();
  const udData = useUDData();
  const udConfig = udData.Config;
  const { basePath, paths } = headerStore.getInstructorPerformancePaths();
  const { gettext } = useI18n();
  function linkSection(text, path, cssToggleId) {
    if (cssToggleId) {
      return /*#__PURE__*/ React.createElement(
        MobileNavItem,
        {
          cssToggleId: cssToggleId,
        },
        /*#__PURE__*/ React.createElement("span", null, text)
      );
    }
    return /*#__PURE__*/ React.createElement(
      MobileNavItem,
      {
        href: `${basePath}${path}`,
      },
      text
    );
  }
  function renderPracticeInsightsMenuItem() {
    return linkSection(
      gettext("Engagement"),
      null,
      "header-toggle-side-nav-instructor-performance-engagement"
    );
  }
  return /*#__PURE__*/ React.createElement(
    MobileNavL1Nav,
    {
      id: "header-toggle-side-nav-instructor-performance",
    },
    /*#__PURE__*/ React.createElement(
      MobileNavSection,
      null,
      !udConfig.brand.has_organization &&
        linkSection(gettext("Overview"), paths.overviewPath),
      !udConfig.brand.has_organization &&
        linkSection(gettext("Students"), paths.studentsPath),
      udConfig.brand.is_add_reviews_enabled &&
        linkSection(gettext("Reviews"), paths.reviewsPath),
      headerStore.isDisplayPracticeInsightsNewPageWithFunnelViewEnabled
        ? renderPracticeInsightsMenuItem()
        : linkSection(gettext("Course engagement"), paths.engagementPath),
      !udConfig.brand.has_organization &&
        linkSection(gettext("Traffic & conversion"), paths.conversionPath)
    )
  );
});
