// @udemy/footer/dist/esm/components/footer/footer.mobx-store.js
import _initializerDefineProperty from "@babel/runtime/helpers/initializerDefineProperty";
import _applyDecoratedDescriptor from "@babel/runtime/helpers/applyDecoratedDescriptor";
import _initializerWarningHelper from "@babel/runtime/helpers/initializerWarningHelper";
var _dec,
  _dec2,
  _dec3,
  _class,
  _descriptor,
  _descriptor2,
  _descriptor3,
  _descriptor4,
  _descriptor5,
  _descriptor6;
import { observable, action } from "mobx";
export let FooterStore =
  ((_dec = observable.ref),
  (_dec2 = observable.ref),
  (_dec3 = observable.ref),
  ((_class = class FooterStore {
    constructor(serverData, clientData) {
      // Server data
      _initializerDefineProperty(
        this,
        "hideFooterUntilContentReady",
        _descriptor,
        this
      );
      _initializerDefineProperty(this, "currentLocaleId", _descriptor2, this);
      _initializerDefineProperty(this, "isJpFooter", _descriptor3, this);
      _initializerDefineProperty(this, "linkColumns", _descriptor4, this);
      _initializerDefineProperty(this, "ufbNotice", _descriptor5, this);
      // Client data
      _initializerDefineProperty(this, "user", _descriptor6, this);
      this.updateServerData(serverData);
      if (clientData) {
        this.updateClientData(clientData);
      }
    }
    updateServerData(data) {
      var _data$hideFooterUntil;
      this.currentLocaleId = data.currentLocaleId;
      this.linkColumns = data.linkColumns;
      this.isJpFooter = data.isJpFooter;
      this.ufbNotice = data.ufbNotice;
      this.hideFooterUntilContentReady =
        (_data$hideFooterUntil =
          data === null || data === void 0
            ? void 0
            : data.hideFooterUntilContentReady) !== null &&
        _data$hideFooterUntil !== void 0
          ? _data$hideFooterUntil
          : false;
    }
    updateClientData(data) {
      this.user = data.user;
    }
  }),
  ((_descriptor = _applyDecoratedDescriptor(
    _class.prototype,
    "hideFooterUntilContentReady",
    [observable],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return false;
      },
    }
  )),
  (_descriptor2 = _applyDecoratedDescriptor(
    _class.prototype,
    "currentLocaleId",
    [observable],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return "en_US";
      },
    }
  )),
  (_descriptor3 = _applyDecoratedDescriptor(
    _class.prototype,
    "isJpFooter",
    [observable],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return false;
      },
    }
  )),
  (_descriptor4 = _applyDecoratedDescriptor(
    _class.prototype,
    "linkColumns",
    [_dec],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return [];
      },
    }
  )),
  (_descriptor5 = _applyDecoratedDescriptor(
    _class.prototype,
    "ufbNotice",
    [_dec2],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return {
          link: undefined,
          placement: undefined,
          isOnsiteRequestDemo: false,
        };
      },
    }
  )),
  (_descriptor6 = _applyDecoratedDescriptor(_class.prototype, "user", [_dec3], {
    configurable: true,
    enumerable: true,
    writable: true,
    initializer: function () {
      return null;
    },
  })),
  _applyDecoratedDescriptor(
    _class.prototype,
    "updateServerData",
    [action],
    Object.getOwnPropertyDescriptor(_class.prototype, "updateServerData"),
    _class.prototype
  ),
  _applyDecoratedDescriptor(
    _class.prototype,
    "updateClientData",
    [action],
    Object.getOwnPropertyDescriptor(_class.prototype, "updateClientData"),
    _class.prototype
  )),
  _class));
