// @udemy/react-header/dist/esm/desktop/instructor/learner-dropdown.react-component.js
import { observer } from "mobx-react";
import React from "react";
import { useI18n } from "@udemy/i18n";
import { useUDData } from "@udemy/ud-data";
import { useHeaderStore } from "../../hooks/use-header-store";
import {
  HeaderDropdown,
  HeaderButton,
  HeaderMenu,
} from "../header-dropdown.react-component";
import styles from "../panel-menu.module.css";
export const LearnerDropdown = observer((_ref) => {
  let { className } = _ref;
  const headerStore = useHeaderStore();
  const { Config } = useUDData();
  const { gettext } = useI18n();
  let buttonText, panelText;
  if (Config.brand.has_organization) {
    buttonText = gettext("Learner");
    panelText = gettext(
      "Switch to the learn view here - get back to the courses you’re taking."
    );
  } else {
    buttonText = gettext("Student");
    panelText = gettext(
      "Switch to the student view here - get back to the courses you’re taking."
    );
  }
  return /*#__PURE__*/ React.createElement(
    HeaderDropdown,
    {
      className: className,
      a11yRole: "description",
      trigger: /*#__PURE__*/ React.createElement(
        HeaderButton,
        {
          componentClass: "a",
          href: headerStore.instructorURLs.BROWSE,
          "data-testid": "instructor-dropdown",
        },
        buttonText
      ),
    },
    /*#__PURE__*/ React.createElement(
      HeaderMenu,
      null,
      /*#__PURE__*/ React.createElement(
        "div",
        {
          className: styles.panel,
          "data-testid": "switch-view-panel",
        },
        panelText
      )
    )
  );
});
