// @udemy/react-header/dist/esm/events.js
import { ClientEvent } from "@udemy/event-tracking";
export class CategoryNavItemSelectEvent extends ClientEvent {
  constructor(_ref) {
    let { context } = _ref;
    super("CategoryNavItemSelectEvent");
    this.categoryId = void 0;
    this.subcategoryId = void 0;
    this.topicId = void 0;
    this.categoryId = context.categoryId;
    this.subcategoryId = context.subcategoryId;
    this.topicId = context.topicId;
  }
}
export let LearningTypeNavItemType = /*#__PURE__*/ (function (
  LearningTypeNavItemType
) {
  LearningTypeNavItemType["LABS"] = "labs";
  LearningTypeNavItemType["ASSESSMENTS"] = "assessments";
  LearningTypeNavItemType["UDEMY_PRO_PATHS"] = "pro_paths";
  return LearningTypeNavItemType;
})({});
export class LearningTypeNavItemSelectEvent extends ClientEvent {
  constructor(itemType) {
    super("LearningTypeNavItemSelectEvent");
    this.itemType = void 0;
    this.itemType = itemType;
  }
}
export let BadgeNavItemType = /*#__PURE__*/ (function (BadgeNavItemType) {
  BadgeNavItemType["CERTIFICATIONS"] = "certifications";
  return BadgeNavItemType;
})({});
export class BadgeNavItemSelectEvent extends ClientEvent {
  constructor(itemType) {
    super("BadgeNavItemSelectEvent");
    this.itemType = void 0;
    this.itemType = itemType;
  }
}
