// @udemy/react-header/dist/esm/logo.react-component.js
import React from "react";
import { Image } from "@udemy/react-core-components";
import { useUDData } from "@udemy/ud-data";
import { udemyLogoUrl } from "./external/udemy-logo-urls";
import styles from "./logo.module.css";
export const Logo = (_ref) => {
  var _ufbContext$logo;
  let { ufbContext, ...imageProps } = _ref;
  const udData = useUDData();
  function renderMarketplaceLogo() {
    return /*#__PURE__*/ React.createElement(
      Image,
      Object.assign(
        {
          src: udemyLogoUrl,
          alt: "Udemy",
          width: 75,
          height: 28,
        },
        imageProps
      )
    );
  }
  if (
    typeof window !== "undefined" &&
    window.location.hostname === "www.udemy.com"
  ) {
    return renderMarketplaceLogo();
  }
  if (udData.isGlobalMeContextLoading) {
    return /*#__PURE__*/ React.createElement("div", {
      className: styles["logo-placeholder"],
    });
  }
  return (_ufbContext$logo =
    ufbContext === null || ufbContext === void 0 ? void 0 : ufbContext.logo) !==
    null && _ufbContext$logo !== void 0
    ? _ufbContext$logo
    : renderMarketplaceLogo();
};
