// @udemy/funnel-tracking/dist/esm/constants.js
export const funnelTrackingConstants = Object.freeze({
  allCoursesContext: "all-courses",
  featuredContext: "featured",
  featuredPageContext: "feature",
  myCoursesContext: "my-courses",
  learningContext: "learning",
  collectionContext: "collection",
  wishlistContext: "wishlist",
  searchPage: "search",
  // for use in subcontext2
  becauseYouEnrolled: "enroll-recommendation",
  becauseYouSearched: "search-recommendation",
  becauseYouViewed: "view-recommendation",
  bestsellers: "best-seller",
  studentsAreViewing: "students-are-viewing",
  newAndNoteworthy: "new-and-noteworthy",
  learningPack: "learning-pack",
  yourWishlist: "wishlist",
  frequentlyBought: "frequentitemsrecommendation",
  tagBasedUnit: "tag-du",
  topicPage: "topic",
  topic: "all-courses",
  onboardingContext: "onboarding-success",
  channelContextMap: {
    category: "cat_ch",
    collection: "coll_ch",
    dynamic_tag_collection: "cl",
    keyword: "kw_ch",
    featured: "ft_ch",
    logged_in_homepage: "ft_ch",
    logged_out_homepage: "home_ch",
    personalized_home: "home_ch",
    home: "home_ch",
    subcategory: "scat_ch",
    topic: "cl",
    landingPageCourse: "landing-page",
    landingPageFreeCourse: "clp-free",
  },
});
