// @udemy/shopping/dist/esm/coupon-store/base-coupon-store.js
import _initializerDefineProperty from "@babel/runtime/helpers/initializerDefineProperty";
import _applyDecoratedDescriptor from "@babel/runtime/helpers/applyDecoratedDescriptor";
import _initializerWarningHelper from "@babel/runtime/helpers/initializerWarningHelper";
var _class,
  _descriptor,
  _descriptor2,
  _descriptor3,
  _descriptor4,
  _descriptor5,
  _descriptor6,
  _descriptor7,
  _descriptor8,
  _descriptor9;
import { action, computed, observable } from "mobx";
import { shoppingConfig } from "../shopping-client/config";
import { MODES } from "./constants";
export let BaseCouponStore =
  ((_class = class BaseCouponStore {
    constructor(i18n) {
      this.i18n = i18n;
      _initializerDefineProperty(this, "currentMode", _descriptor, this);
      _initializerDefineProperty(this, "inputCode", _descriptor2, this);
      _initializerDefineProperty(this, "submittedCode", _descriptor3, this);
      _initializerDefineProperty(this, "error", _descriptor4, this);
      _initializerDefineProperty(this, "currentCodes", _descriptor5, this);
      _initializerDefineProperty(this, "currentDiscounts", _descriptor6, this);
      _initializerDefineProperty(
        this,
        "hasAlreadyPurchased",
        _descriptor7,
        this
      );
      _initializerDefineProperty(this, "enterInputMode", _descriptor8, this);
      _initializerDefineProperty(this, "enterLoadingMode", _descriptor9, this);
    }
    get isValidInputCode() {
      return shoppingConfig.patterns.validDiscountCode.test(this.inputCode);
    }
    _updateBaseCouponInternalState(lastDiscountAttempt) {
      this.submittedCode = this.inputCode;
      const errors = shoppingConfig.errors(this.i18n.gettext);
      this.error = errors.discountCodeInputFormat;
      this.currentMode = MODES.INPUT;
      this.inputCode = "";
      if (lastDiscountAttempt) {
        this.error = lastDiscountAttempt.length
          ? lastDiscountAttempt[0].details
          : "";
      }
    }
    setCode(code) {
      this.inputCode = code.trim().toUpperCase();
      if (this.error) {
        this.error = "";
      }
      if (this.submittedCode) {
        this.submittedCode = "";
      }
    }
    get allowSubmit() {
      return (
        this.currentMode === MODES.INPUT &&
        !this.currentCodes.includes(this.inputCode)
      );
    }
  }),
  ((_descriptor = _applyDecoratedDescriptor(
    _class.prototype,
    "currentMode",
    [observable],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return MODES.OFF;
      },
    }
  )),
  (_descriptor2 = _applyDecoratedDescriptor(
    _class.prototype,
    "inputCode",
    [observable],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return "";
      },
    }
  )),
  (_descriptor3 = _applyDecoratedDescriptor(
    _class.prototype,
    "submittedCode",
    [observable],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return "";
      },
    }
  )),
  (_descriptor4 = _applyDecoratedDescriptor(
    _class.prototype,
    "error",
    [observable],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return "";
      },
    }
  )),
  (_descriptor5 = _applyDecoratedDescriptor(
    _class.prototype,
    "currentCodes",
    [observable],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return [];
      },
    }
  )),
  (_descriptor6 = _applyDecoratedDescriptor(
    _class.prototype,
    "currentDiscounts",
    [observable],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return new Map();
      },
    }
  )),
  (_descriptor7 = _applyDecoratedDescriptor(
    _class.prototype,
    "hasAlreadyPurchased",
    [observable],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return false;
      },
    }
  )),
  (_descriptor8 = _applyDecoratedDescriptor(
    _class.prototype,
    "enterInputMode",
    [action],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return () => {
          this.currentMode = MODES.INPUT;
        };
      },
    }
  )),
  (_descriptor9 = _applyDecoratedDescriptor(
    _class.prototype,
    "enterLoadingMode",
    [action],
    {
      configurable: true,
      enumerable: true,
      writable: true,
      initializer: function () {
        return () => {
          this.currentMode = MODES.PENDING;
        };
      },
    }
  )),
  _applyDecoratedDescriptor(
    _class.prototype,
    "isValidInputCode",
    [computed],
    Object.getOwnPropertyDescriptor(_class.prototype, "isValidInputCode"),
    _class.prototype
  ),
  _applyDecoratedDescriptor(
    _class.prototype,
    "_updateBaseCouponInternalState",
    [action],
    Object.getOwnPropertyDescriptor(
      _class.prototype,
      "_updateBaseCouponInternalState"
    ),
    _class.prototype
  ),
  _applyDecoratedDescriptor(
    _class.prototype,
    "setCode",
    [action],
    Object.getOwnPropertyDescriptor(_class.prototype, "setCode"),
    _class.prototype
  ),
  _applyDecoratedDescriptor(
    _class.prototype,
    "allowSubmit",
    [computed],
    Object.getOwnPropertyDescriptor(_class.prototype, "allowSubmit"),
    _class.prototype
  )),
  _class);
