// @udemy/react-header/dist/esm/desktop/try-ufb/try-ufb-dropdown.react-component.js
import { observer } from "mobx-react";
import React from "react";
import { UFBNoticeClickEvent } from "@udemy/browse-event-tracking";
import { Tracker } from "@udemy/event-tracking";
import { useI18n } from "@udemy/i18n";
import { Button } from "@udemy/react-core-components";
import { useSiteStats, useUDData } from "@udemy/ud-data";
import { useHeaderStore } from "../../hooks/use-header-store";
import {
  HeaderButton,
  HeaderDropdown,
  HeaderMenu,
} from "../header-dropdown.react-component";
import styles from "../panel-menu.module.css";
// TODO: Add UFB check here, and on dropdown
export const TryUFBDropdown = observer((_ref) => {
  let { className } = _ref;
  const headerStore = useHeaderStore();
  const { Config } = useUDData();
  const { gettext, interpolate, locale } = useI18n();
  const siteStats = useSiteStats();
  const handleClick = () => {
    if (headerStore.tryUFBPlacements) {
      var _headerStore$tryUFBPl;
      Tracker.publishEvent(
        new UFBNoticeClickEvent({
          locale,
          placement:
            (_headerStore$tryUFBPl = headerStore.tryUFBPlacements) === null ||
            _headerStore$tryUFBPl === void 0
              ? void 0
              : _headerStore$tryUFBPl.bar,
          variant: headerStore.userSpecificContext.ufbHookVariant,
          url: headerStore.userSpecificContext.productLink.url,
        })
      );
    }
  };
  const copyText = interpolate(
    gettext(
      "Get your team access to over %(count)s top Udemy courses, anytime, anywhere."
    ),
    {
      count: siteStats.getOrgNumericSiteStat("num_courses"),
    },
    true
  );
  const ctaText = interpolate(
    gettext("Try %(product)s"),
    {
      product: headerStore.userSpecificContext.productLink.text,
    },
    true
  );
  const { productLink } = headerStore.userSpecificContext;
  const hasOrganization = Config.brand.has_organization;
  return !hasOrganization
    ? /*#__PURE__*/ React.createElement(
        HeaderDropdown,
        {
          className: className,
          a11yRole: "description",
          trigger: /*#__PURE__*/ React.createElement(
            HeaderButton,
            {
              componentClass: "a",
              "data-purpose": "try-ufb-dropdown-trigger",
              href: productLink.url,
              target: productLink.isOnsiteRequestDemo ? undefined : "_blank",
              rel: productLink.isOnsiteRequestDemo ? undefined : "noopener",
              onClick: handleClick,
            },
            gettext("Udemy Business")
          ),
        },
        /*#__PURE__*/ React.createElement(
          HeaderMenu,
          null,
          /*#__PURE__*/ React.createElement(
            "div",
            {
              className: styles["panel"],
            },
            /*#__PURE__*/ React.createElement(
              "div",
              {
                className: `ud-heading-lg ${styles["gap-bottom"]}`,
                "data-purpose": "hook-text",
              },
              copyText
            ),
            /*#__PURE__*/ React.createElement(
              Button,
              {
                componentClass: "a",
                "data-purpose": "try-ufb-button",
                href: productLink.url,
                className: styles["cta"],
                target: productLink.isOnsiteRequestDemo ? undefined : "_blank",
                rel: productLink.isOnsiteRequestDemo ? undefined : "noopener",
                onClick: handleClick,
              },
              ctaText
            )
          )
        )
      )
    : null;
});
