// @udemy/protobuf-objects/build/protos/udemy/dto/service_request_context/v1beta/bot_information_pb.js
// source: udemy/dto/service_request_context/v1beta/bot_information.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require("google-protobuf");
var goog = jspb;
var global = Function("return this")();

goog.exportSymbol(
  "proto.udemy.dto.service_request_context.v1.BotInformation",
  null,
  global
);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.udemy.dto.service_request_context.v1.BotInformation = function (
  opt_data
) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(
  proto.udemy.dto.service_request_context.v1.BotInformation,
  jspb.Message
);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.udemy.dto.service_request_context.v1.BotInformation.displayName =
    "proto.udemy.dto.service_request_context.v1.BotInformation";
}

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * Optional fields that are not set will be set to undefined.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
   * @param {boolean=} opt_includeInstance Deprecated. whether to include the
   *     JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.udemy.dto.service_request_context.v1.BotInformation.prototype.toObject =
    function (opt_includeInstance) {
      return proto.udemy.dto.service_request_context.v1.BotInformation.toObject(
        opt_includeInstance,
        this
      );
    };

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Deprecated. Whether to include
   *     the JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.udemy.dto.service_request_context.v1.BotInformation} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.udemy.dto.service_request_context.v1.BotInformation.toObject =
    function (includeInstance, msg) {
      var f,
        obj = {
          botScore: jspb.Message.getFieldWithDefault(msg, 1, 0),
          isVerifiedBot: jspb.Message.getBooleanFieldWithDefault(msg, 2, false),
        };

      if (includeInstance) {
        obj.$jspbMessageInstance = msg;
      }
      return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.udemy.dto.service_request_context.v1.BotInformation}
 */
proto.udemy.dto.service_request_context.v1.BotInformation.deserializeBinary =
  function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.udemy.dto.service_request_context.v1.BotInformation();
    return proto.udemy.dto.service_request_context.v1.BotInformation.deserializeBinaryFromReader(
      msg,
      reader
    );
  };

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.udemy.dto.service_request_context.v1.BotInformation} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.udemy.dto.service_request_context.v1.BotInformation}
 */
proto.udemy.dto.service_request_context.v1.BotInformation.deserializeBinaryFromReader =
  function (msg, reader) {
    while (reader.nextField()) {
      if (reader.isEndGroup()) {
        break;
      }
      var field = reader.getFieldNumber();
      switch (field) {
        case 1:
          var value = /** @type {number} */ (reader.readUint32());
          msg.setBotScore(value);
          break;
        case 2:
          var value = /** @type {boolean} */ (reader.readBool());
          msg.setIsVerifiedBot(value);
          break;
        default:
          reader.skipField();
          break;
      }
    }
    return msg;
  };

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.udemy.dto.service_request_context.v1.BotInformation.prototype.serializeBinary =
  function () {
    var writer = new jspb.BinaryWriter();
    proto.udemy.dto.service_request_context.v1.BotInformation.serializeBinaryToWriter(
      this,
      writer
    );
    return writer.getResultBuffer();
  };

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.udemy.dto.service_request_context.v1.BotInformation} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.udemy.dto.service_request_context.v1.BotInformation.serializeBinaryToWriter =
  function (message, writer) {
    var f = undefined;
    f = message.getBotScore();
    if (f !== 0) {
      writer.writeUint32(1, f);
    }
    f = message.getIsVerifiedBot();
    if (f) {
      writer.writeBool(2, f);
    }
  };

/**
 * optional uint32 bot_score = 1;
 * @return {number}
 */
proto.udemy.dto.service_request_context.v1.BotInformation.prototype.getBotScore =
  function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 1, 0));
  };

/**
 * @param {number} value
 * @return {!proto.udemy.dto.service_request_context.v1.BotInformation} returns this
 */
proto.udemy.dto.service_request_context.v1.BotInformation.prototype.setBotScore =
  function (value) {
    return jspb.Message.setProto3IntField(this, 1, value);
  };

/**
 * optional bool is_verified_bot = 2;
 * @return {boolean}
 */
proto.udemy.dto.service_request_context.v1.BotInformation.prototype.getIsVerifiedBot =
  function () {
    return /** @type {boolean} */ (
      jspb.Message.getBooleanFieldWithDefault(this, 2, false)
    );
  };

/**
 * @param {boolean} value
 * @return {!proto.udemy.dto.service_request_context.v1.BotInformation} returns this
 */
proto.udemy.dto.service_request_context.v1.BotInformation.prototype.setIsVerifiedBot =
  function (value) {
    return jspb.Message.setProto3BooleanField(this, 2, value);
  };

goog.object.extend(exports, proto.udemy.dto.service_request_context.v1);
