// @udemy/protobuf-objects/build/protos/udemy/dto/service_request_context/v1beta/request_context_attributions_pb.js
// source: udemy/dto/service_request_context/v1beta/request_context_attributions.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require("google-protobuf");
var goog = jspb;
var global = Function("return this")();

var udemy_dto_service_request_context_v1beta_bot_information_pb = require("../../../../udemy/dto/service_request_context/v1beta/bot_information_pb.js");
goog.object.extend(
  proto,
  udemy_dto_service_request_context_v1beta_bot_information_pb
);
goog.exportSymbol(
  "proto.udemy.dto.service_request_context.v1.RequestContextAttributions",
  null,
  global
);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.udemy.dto.service_request_context.v1.RequestContextAttributions =
  function (opt_data) {
    jspb.Message.initialize(this, opt_data, 0, -1, null, null);
  };
goog.inherits(
  proto.udemy.dto.service_request_context.v1.RequestContextAttributions,
  jspb.Message
);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.udemy.dto.service_request_context.v1.RequestContextAttributions.displayName =
    "proto.udemy.dto.service_request_context.v1.RequestContextAttributions";
}

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * Optional fields that are not set will be set to undefined.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
   * @param {boolean=} opt_includeInstance Deprecated. whether to include the
   *     JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.udemy.dto.service_request_context.v1.RequestContextAttributions.prototype.toObject =
    function (opt_includeInstance) {
      return proto.udemy.dto.service_request_context.v1.RequestContextAttributions.toObject(
        opt_includeInstance,
        this
      );
    };

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Deprecated. Whether to include
   *     the JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.udemy.dto.service_request_context.v1.RequestContextAttributions} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.udemy.dto.service_request_context.v1.RequestContextAttributions.toObject =
    function (includeInstance, msg) {
      var f,
        obj = {
          visitorUuid: jspb.Message.getFieldWithDefault(msg, 1, ""),
          botInformation:
            (f = msg.getBotInformation()) &&
            udemy_dto_service_request_context_v1beta_bot_information_pb.BotInformation.toObject(
              includeInstance,
              f
            ),
        };

      if (includeInstance) {
        obj.$jspbMessageInstance = msg;
      }
      return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.udemy.dto.service_request_context.v1.RequestContextAttributions}
 */
proto.udemy.dto.service_request_context.v1.RequestContextAttributions.deserializeBinary =
  function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg =
      new proto.udemy.dto.service_request_context.v1.RequestContextAttributions();
    return proto.udemy.dto.service_request_context.v1.RequestContextAttributions.deserializeBinaryFromReader(
      msg,
      reader
    );
  };

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.udemy.dto.service_request_context.v1.RequestContextAttributions} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.udemy.dto.service_request_context.v1.RequestContextAttributions}
 */
proto.udemy.dto.service_request_context.v1.RequestContextAttributions.deserializeBinaryFromReader =
  function (msg, reader) {
    while (reader.nextField()) {
      if (reader.isEndGroup()) {
        break;
      }
      var field = reader.getFieldNumber();
      switch (field) {
        case 1:
          var value = /** @type {string} */ (reader.readString());
          msg.setVisitorUuid(value);
          break;
        case 2:
          var value =
            new udemy_dto_service_request_context_v1beta_bot_information_pb.BotInformation();
          reader.readMessage(
            value,
            udemy_dto_service_request_context_v1beta_bot_information_pb
              .BotInformation.deserializeBinaryFromReader
          );
          msg.setBotInformation(value);
          break;
        default:
          reader.skipField();
          break;
      }
    }
    return msg;
  };

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.udemy.dto.service_request_context.v1.RequestContextAttributions.prototype.serializeBinary =
  function () {
    var writer = new jspb.BinaryWriter();
    proto.udemy.dto.service_request_context.v1.RequestContextAttributions.serializeBinaryToWriter(
      this,
      writer
    );
    return writer.getResultBuffer();
  };

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.udemy.dto.service_request_context.v1.RequestContextAttributions} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.udemy.dto.service_request_context.v1.RequestContextAttributions.serializeBinaryToWriter =
  function (message, writer) {
    var f = undefined;
    f = message.getVisitorUuid();
    if (f.length > 0) {
      writer.writeString(1, f);
    }
    f = message.getBotInformation();
    if (f != null) {
      writer.writeMessage(
        2,
        f,
        udemy_dto_service_request_context_v1beta_bot_information_pb
          .BotInformation.serializeBinaryToWriter
      );
    }
  };

/**
 * optional string visitor_uuid = 1;
 * @return {string}
 */
proto.udemy.dto.service_request_context.v1.RequestContextAttributions.prototype.getVisitorUuid =
  function () {
    return /** @type {string} */ (
      jspb.Message.getFieldWithDefault(this, 1, "")
    );
  };

/**
 * @param {string} value
 * @return {!proto.udemy.dto.service_request_context.v1.RequestContextAttributions} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestContextAttributions.prototype.setVisitorUuid =
  function (value) {
    return jspb.Message.setProto3StringField(this, 1, value);
  };

/**
 * optional BotInformation bot_information = 2;
 * @return {?proto.udemy.dto.service_request_context.v1.BotInformation}
 */
proto.udemy.dto.service_request_context.v1.RequestContextAttributions.prototype.getBotInformation =
  function () {
    return /** @type{?proto.udemy.dto.service_request_context.v1.BotInformation} */ (
      jspb.Message.getWrapperField(
        this,
        udemy_dto_service_request_context_v1beta_bot_information_pb.BotInformation,
        2
      )
    );
  };

/**
 * @param {?proto.udemy.dto.service_request_context.v1.BotInformation|undefined} value
 * @return {!proto.udemy.dto.service_request_context.v1.RequestContextAttributions} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestContextAttributions.prototype.setBotInformation =
  function (value) {
    return jspb.Message.setWrapperField(this, 2, value);
  };

/**
 * Clears the message field making it undefined.
 * @return {!proto.udemy.dto.service_request_context.v1.RequestContextAttributions} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestContextAttributions.prototype.clearBotInformation =
  function () {
    return this.setBotInformation(undefined);
  };

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.udemy.dto.service_request_context.v1.RequestContextAttributions.prototype.hasBotInformation =
  function () {
    return jspb.Message.getField(this, 2) != null;
  };

goog.object.extend(exports, proto.udemy.dto.service_request_context.v1);
