// @udemy/protobuf-objects/build/protos/udemy/dto/service_request_context/v1beta/request_origin_information_pb.js
// source: udemy/dto/service_request_context/v1beta/request_origin_information.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require("google-protobuf");
var goog = jspb;
var global = Function("return this")();

var udemy_dto_service_request_context_v1beta_application_information_pb = require("../../../../udemy/dto/service_request_context/v1beta/application_information_pb.js");
goog.object.extend(
  proto,
  udemy_dto_service_request_context_v1beta_application_information_pb
);
var udemy_dto_service_request_context_v1beta_location_information_pb = require("../../../../udemy/dto/service_request_context/v1beta/location_information_pb.js");
goog.object.extend(
  proto,
  udemy_dto_service_request_context_v1beta_location_information_pb
);
goog.exportSymbol(
  "proto.udemy.dto.service_request_context.v1.RequestOriginInformation",
  null,
  global
);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation = function (
  opt_data
) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(
  proto.udemy.dto.service_request_context.v1.RequestOriginInformation,
  jspb.Message
);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.udemy.dto.service_request_context.v1.RequestOriginInformation.displayName =
    "proto.udemy.dto.service_request_context.v1.RequestOriginInformation";
}

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * Optional fields that are not set will be set to undefined.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
   * @param {boolean=} opt_includeInstance Deprecated. whether to include the
   *     JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.toObject =
    function (opt_includeInstance) {
      return proto.udemy.dto.service_request_context.v1.RequestOriginInformation.toObject(
        opt_includeInstance,
        this
      );
    };

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Deprecated. Whether to include
   *     the JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.udemy.dto.service_request_context.v1.RequestOriginInformation.toObject =
    function (includeInstance, msg) {
      var f,
        obj = {
          requestStartTime: jspb.Message.getFieldWithDefault(msg, 1, 0),
          host: jspb.Message.getFieldWithDefault(msg, 2, ""),
          userAgent: jspb.Message.getFieldWithDefault(msg, 3, ""),
          ipCountry: jspb.Message.getFieldWithDefault(msg, 4, ""),
          detectedDevice: jspb.Message.getFieldWithDefault(msg, 5, ""),
          locationInformation:
            (f = msg.getLocationInformation()) &&
            udemy_dto_service_request_context_v1beta_location_information_pb.LocationInformation.toObject(
              includeInstance,
              f
            ),
          applicationInformation:
            (f = msg.getApplicationInformation()) &&
            udemy_dto_service_request_context_v1beta_application_information_pb.ApplicationInformation.toObject(
              includeInstance,
              f
            ),
          acceptLanguage: jspb.Message.getFieldWithDefault(msg, 8, ""),
          ipAddress: jspb.Message.getFieldWithDefault(msg, 9, ""),
        };

      if (includeInstance) {
        obj.$jspbMessageInstance = msg;
      }
      return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.deserializeBinary =
  function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg =
      new proto.udemy.dto.service_request_context.v1.RequestOriginInformation();
    return proto.udemy.dto.service_request_context.v1.RequestOriginInformation.deserializeBinaryFromReader(
      msg,
      reader
    );
  };

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.deserializeBinaryFromReader =
  function (msg, reader) {
    while (reader.nextField()) {
      if (reader.isEndGroup()) {
        break;
      }
      var field = reader.getFieldNumber();
      switch (field) {
        case 1:
          var value = /** @type {number} */ (reader.readUint64());
          msg.setRequestStartTime(value);
          break;
        case 2:
          var value = /** @type {string} */ (reader.readString());
          msg.setHost(value);
          break;
        case 3:
          var value = /** @type {string} */ (reader.readString());
          msg.setUserAgent(value);
          break;
        case 4:
          var value = /** @type {string} */ (reader.readString());
          msg.setIpCountry(value);
          break;
        case 5:
          var value = /** @type {string} */ (reader.readString());
          msg.setDetectedDevice(value);
          break;
        case 6:
          var value =
            new udemy_dto_service_request_context_v1beta_location_information_pb.LocationInformation();
          reader.readMessage(
            value,
            udemy_dto_service_request_context_v1beta_location_information_pb
              .LocationInformation.deserializeBinaryFromReader
          );
          msg.setLocationInformation(value);
          break;
        case 7:
          var value =
            new udemy_dto_service_request_context_v1beta_application_information_pb.ApplicationInformation();
          reader.readMessage(
            value,
            udemy_dto_service_request_context_v1beta_application_information_pb
              .ApplicationInformation.deserializeBinaryFromReader
          );
          msg.setApplicationInformation(value);
          break;
        case 8:
          var value = /** @type {string} */ (reader.readString());
          msg.setAcceptLanguage(value);
          break;
        case 9:
          var value = /** @type {string} */ (reader.readString());
          msg.setIpAddress(value);
          break;
        default:
          reader.skipField();
          break;
      }
    }
    return msg;
  };

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.serializeBinary =
  function () {
    var writer = new jspb.BinaryWriter();
    proto.udemy.dto.service_request_context.v1.RequestOriginInformation.serializeBinaryToWriter(
      this,
      writer
    );
    return writer.getResultBuffer();
  };

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.serializeBinaryToWriter =
  function (message, writer) {
    var f = undefined;
    f = message.getRequestStartTime();
    if (f !== 0) {
      writer.writeUint64(1, f);
    }
    f = message.getHost();
    if (f.length > 0) {
      writer.writeString(2, f);
    }
    f = message.getUserAgent();
    if (f.length > 0) {
      writer.writeString(3, f);
    }
    f = message.getIpCountry();
    if (f.length > 0) {
      writer.writeString(4, f);
    }
    f = message.getDetectedDevice();
    if (f.length > 0) {
      writer.writeString(5, f);
    }
    f = message.getLocationInformation();
    if (f != null) {
      writer.writeMessage(
        6,
        f,
        udemy_dto_service_request_context_v1beta_location_information_pb
          .LocationInformation.serializeBinaryToWriter
      );
    }
    f = message.getApplicationInformation();
    if (f != null) {
      writer.writeMessage(
        7,
        f,
        udemy_dto_service_request_context_v1beta_application_information_pb
          .ApplicationInformation.serializeBinaryToWriter
      );
    }
    f = message.getAcceptLanguage();
    if (f.length > 0) {
      writer.writeString(8, f);
    }
    f = message.getIpAddress();
    if (f.length > 0) {
      writer.writeString(9, f);
    }
  };

/**
 * optional uint64 request_start_time = 1;
 * @return {number}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.getRequestStartTime =
  function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 1, 0));
  };

/**
 * @param {number} value
 * @return {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.setRequestStartTime =
  function (value) {
    return jspb.Message.setProto3IntField(this, 1, value);
  };

/**
 * optional string host = 2;
 * @return {string}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.getHost =
  function () {
    return /** @type {string} */ (
      jspb.Message.getFieldWithDefault(this, 2, "")
    );
  };

/**
 * @param {string} value
 * @return {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.setHost =
  function (value) {
    return jspb.Message.setProto3StringField(this, 2, value);
  };

/**
 * optional string user_agent = 3;
 * @return {string}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.getUserAgent =
  function () {
    return /** @type {string} */ (
      jspb.Message.getFieldWithDefault(this, 3, "")
    );
  };

/**
 * @param {string} value
 * @return {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.setUserAgent =
  function (value) {
    return jspb.Message.setProto3StringField(this, 3, value);
  };

/**
 * optional string ip_country = 4;
 * @return {string}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.getIpCountry =
  function () {
    return /** @type {string} */ (
      jspb.Message.getFieldWithDefault(this, 4, "")
    );
  };

/**
 * @param {string} value
 * @return {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.setIpCountry =
  function (value) {
    return jspb.Message.setProto3StringField(this, 4, value);
  };

/**
 * optional string detected_device = 5;
 * @return {string}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.getDetectedDevice =
  function () {
    return /** @type {string} */ (
      jspb.Message.getFieldWithDefault(this, 5, "")
    );
  };

/**
 * @param {string} value
 * @return {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.setDetectedDevice =
  function (value) {
    return jspb.Message.setProto3StringField(this, 5, value);
  };

/**
 * optional LocationInformation location_information = 6;
 * @return {?proto.udemy.dto.service_request_context.v1.LocationInformation}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.getLocationInformation =
  function () {
    return /** @type{?proto.udemy.dto.service_request_context.v1.LocationInformation} */ (
      jspb.Message.getWrapperField(
        this,
        udemy_dto_service_request_context_v1beta_location_information_pb.LocationInformation,
        6
      )
    );
  };

/**
 * @param {?proto.udemy.dto.service_request_context.v1.LocationInformation|undefined} value
 * @return {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.setLocationInformation =
  function (value) {
    return jspb.Message.setWrapperField(this, 6, value);
  };

/**
 * Clears the message field making it undefined.
 * @return {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.clearLocationInformation =
  function () {
    return this.setLocationInformation(undefined);
  };

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.hasLocationInformation =
  function () {
    return jspb.Message.getField(this, 6) != null;
  };

/**
 * optional ApplicationInformation application_information = 7;
 * @return {?proto.udemy.dto.service_request_context.v1.ApplicationInformation}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.getApplicationInformation =
  function () {
    return /** @type{?proto.udemy.dto.service_request_context.v1.ApplicationInformation} */ (
      jspb.Message.getWrapperField(
        this,
        udemy_dto_service_request_context_v1beta_application_information_pb.ApplicationInformation,
        7
      )
    );
  };

/**
 * @param {?proto.udemy.dto.service_request_context.v1.ApplicationInformation|undefined} value
 * @return {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.setApplicationInformation =
  function (value) {
    return jspb.Message.setWrapperField(this, 7, value);
  };

/**
 * Clears the message field making it undefined.
 * @return {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.clearApplicationInformation =
  function () {
    return this.setApplicationInformation(undefined);
  };

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.hasApplicationInformation =
  function () {
    return jspb.Message.getField(this, 7) != null;
  };

/**
 * optional string accept_language = 8;
 * @return {string}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.getAcceptLanguage =
  function () {
    return /** @type {string} */ (
      jspb.Message.getFieldWithDefault(this, 8, "")
    );
  };

/**
 * @param {string} value
 * @return {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.setAcceptLanguage =
  function (value) {
    return jspb.Message.setProto3StringField(this, 8, value);
  };

/**
 * optional string ip_address = 9;
 * @return {string}
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.getIpAddress =
  function () {
    return /** @type {string} */ (
      jspb.Message.getFieldWithDefault(this, 9, "")
    );
  };

/**
 * @param {string} value
 * @return {!proto.udemy.dto.service_request_context.v1.RequestOriginInformation} returns this
 */
proto.udemy.dto.service_request_context.v1.RequestOriginInformation.prototype.setIpAddress =
  function (value) {
    return jspb.Message.setProto3StringField(this, 9, value);
  };

goog.object.extend(exports, proto.udemy.dto.service_request_context.v1);
