// @udemy/protobuf-objects/build/protos/udemy/dto/service_request_context/v1beta/service_request_context_pb.js
// source: udemy/dto/service_request_context/v1beta/service_request_context.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require("google-protobuf");
var goog = jspb;
var global = Function("return this")();

var udemy_dto_user_v1beta_user_pb = require("../../../../udemy/dto/user/v1beta/user_pb.js");
goog.object.extend(proto, udemy_dto_user_v1beta_user_pb);
var udemy_dto_service_request_context_v1beta_request_origin_information_pb = require("../../../../udemy/dto/service_request_context/v1beta/request_origin_information_pb.js");
goog.object.extend(
  proto,
  udemy_dto_service_request_context_v1beta_request_origin_information_pb
);
var udemy_dto_organization_v1beta_organization_pb = require("../../../../udemy/dto/organization/v1beta/organization_pb.js");
goog.object.extend(proto, udemy_dto_organization_v1beta_organization_pb);
var udemy_dto_service_request_context_v1beta_request_context_attributions_pb = require("../../../../udemy/dto/service_request_context/v1beta/request_context_attributions_pb.js");
goog.object.extend(
  proto,
  udemy_dto_service_request_context_v1beta_request_context_attributions_pb
);
goog.exportSymbol(
  "proto.udemy.dto.service_request_context.v1.ServiceRequestContext",
  null,
  global
);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext = function (
  opt_data
) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(
  proto.udemy.dto.service_request_context.v1.ServiceRequestContext,
  jspb.Message
);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.udemy.dto.service_request_context.v1.ServiceRequestContext.displayName =
    "proto.udemy.dto.service_request_context.v1.ServiceRequestContext";
}

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * Optional fields that are not set will be set to undefined.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
   * @param {boolean=} opt_includeInstance Deprecated. whether to include the
   *     JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.toObject =
    function (opt_includeInstance) {
      return proto.udemy.dto.service_request_context.v1.ServiceRequestContext.toObject(
        opt_includeInstance,
        this
      );
    };

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Deprecated. Whether to include
   *     the JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.udemy.dto.service_request_context.v1.ServiceRequestContext} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.udemy.dto.service_request_context.v1.ServiceRequestContext.toObject =
    function (includeInstance, msg) {
      var f,
        obj = {
          user:
            (f = msg.getUser()) &&
            udemy_dto_user_v1beta_user_pb.User.toObject(includeInstance, f),
          requestOriginInfo:
            (f = msg.getRequestOriginInfo()) &&
            udemy_dto_service_request_context_v1beta_request_origin_information_pb.RequestOriginInformation.toObject(
              includeInstance,
              f
            ),
          organization:
            (f = msg.getOrganization()) &&
            udemy_dto_organization_v1beta_organization_pb.Organization.toObject(
              includeInstance,
              f
            ),
          attributions:
            (f = msg.getAttributions()) &&
            udemy_dto_service_request_context_v1beta_request_context_attributions_pb.RequestContextAttributions.toObject(
              includeInstance,
              f
            ),
        };

      if (includeInstance) {
        obj.$jspbMessageInstance = msg;
      }
      return obj;
    };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.udemy.dto.service_request_context.v1.ServiceRequestContext}
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.deserializeBinary =
  function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg =
      new proto.udemy.dto.service_request_context.v1.ServiceRequestContext();
    return proto.udemy.dto.service_request_context.v1.ServiceRequestContext.deserializeBinaryFromReader(
      msg,
      reader
    );
  };

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.udemy.dto.service_request_context.v1.ServiceRequestContext} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.udemy.dto.service_request_context.v1.ServiceRequestContext}
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.deserializeBinaryFromReader =
  function (msg, reader) {
    while (reader.nextField()) {
      if (reader.isEndGroup()) {
        break;
      }
      var field = reader.getFieldNumber();
      switch (field) {
        case 1:
          var value = new udemy_dto_user_v1beta_user_pb.User();
          reader.readMessage(
            value,
            udemy_dto_user_v1beta_user_pb.User.deserializeBinaryFromReader
          );
          msg.setUser(value);
          break;
        case 2:
          var value =
            new udemy_dto_service_request_context_v1beta_request_origin_information_pb.RequestOriginInformation();
          reader.readMessage(
            value,
            udemy_dto_service_request_context_v1beta_request_origin_information_pb
              .RequestOriginInformation.deserializeBinaryFromReader
          );
          msg.setRequestOriginInfo(value);
          break;
        case 3:
          var value =
            new udemy_dto_organization_v1beta_organization_pb.Organization();
          reader.readMessage(
            value,
            udemy_dto_organization_v1beta_organization_pb.Organization
              .deserializeBinaryFromReader
          );
          msg.setOrganization(value);
          break;
        case 4:
          var value =
            new udemy_dto_service_request_context_v1beta_request_context_attributions_pb.RequestContextAttributions();
          reader.readMessage(
            value,
            udemy_dto_service_request_context_v1beta_request_context_attributions_pb
              .RequestContextAttributions.deserializeBinaryFromReader
          );
          msg.setAttributions(value);
          break;
        default:
          reader.skipField();
          break;
      }
    }
    return msg;
  };

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.serializeBinary =
  function () {
    var writer = new jspb.BinaryWriter();
    proto.udemy.dto.service_request_context.v1.ServiceRequestContext.serializeBinaryToWriter(
      this,
      writer
    );
    return writer.getResultBuffer();
  };

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.udemy.dto.service_request_context.v1.ServiceRequestContext} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.serializeBinaryToWriter =
  function (message, writer) {
    var f = undefined;
    f = message.getUser();
    if (f != null) {
      writer.writeMessage(
        1,
        f,
        udemy_dto_user_v1beta_user_pb.User.serializeBinaryToWriter
      );
    }
    f = message.getRequestOriginInfo();
    if (f != null) {
      writer.writeMessage(
        2,
        f,
        udemy_dto_service_request_context_v1beta_request_origin_information_pb
          .RequestOriginInformation.serializeBinaryToWriter
      );
    }
    f = message.getOrganization();
    if (f != null) {
      writer.writeMessage(
        3,
        f,
        udemy_dto_organization_v1beta_organization_pb.Organization
          .serializeBinaryToWriter
      );
    }
    f = message.getAttributions();
    if (f != null) {
      writer.writeMessage(
        4,
        f,
        udemy_dto_service_request_context_v1beta_request_context_attributions_pb
          .RequestContextAttributions.serializeBinaryToWriter
      );
    }
  };

/**
 * optional udemy.dto.user.v1.User user = 1;
 * @return {?proto.udemy.dto.user.v1.User}
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.getUser =
  function () {
    return /** @type{?proto.udemy.dto.user.v1.User} */ (
      jspb.Message.getWrapperField(this, udemy_dto_user_v1beta_user_pb.User, 1)
    );
  };

/**
 * @param {?proto.udemy.dto.user.v1.User|undefined} value
 * @return {!proto.udemy.dto.service_request_context.v1.ServiceRequestContext} returns this
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.setUser =
  function (value) {
    return jspb.Message.setWrapperField(this, 1, value);
  };

/**
 * Clears the message field making it undefined.
 * @return {!proto.udemy.dto.service_request_context.v1.ServiceRequestContext} returns this
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.clearUser =
  function () {
    return this.setUser(undefined);
  };

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.hasUser =
  function () {
    return jspb.Message.getField(this, 1) != null;
  };

/**
 * optional RequestOriginInformation request_origin_info = 2;
 * @return {?proto.udemy.dto.service_request_context.v1.RequestOriginInformation}
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.getRequestOriginInfo =
  function () {
    return /** @type{?proto.udemy.dto.service_request_context.v1.RequestOriginInformation} */ (
      jspb.Message.getWrapperField(
        this,
        udemy_dto_service_request_context_v1beta_request_origin_information_pb.RequestOriginInformation,
        2
      )
    );
  };

/**
 * @param {?proto.udemy.dto.service_request_context.v1.RequestOriginInformation|undefined} value
 * @return {!proto.udemy.dto.service_request_context.v1.ServiceRequestContext} returns this
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.setRequestOriginInfo =
  function (value) {
    return jspb.Message.setWrapperField(this, 2, value);
  };

/**
 * Clears the message field making it undefined.
 * @return {!proto.udemy.dto.service_request_context.v1.ServiceRequestContext} returns this
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.clearRequestOriginInfo =
  function () {
    return this.setRequestOriginInfo(undefined);
  };

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.hasRequestOriginInfo =
  function () {
    return jspb.Message.getField(this, 2) != null;
  };

/**
 * optional udemy.dto.organization.v1.Organization organization = 3;
 * @return {?proto.udemy.dto.organization.v1.Organization}
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.getOrganization =
  function () {
    return /** @type{?proto.udemy.dto.organization.v1.Organization} */ (
      jspb.Message.getWrapperField(
        this,
        udemy_dto_organization_v1beta_organization_pb.Organization,
        3
      )
    );
  };

/**
 * @param {?proto.udemy.dto.organization.v1.Organization|undefined} value
 * @return {!proto.udemy.dto.service_request_context.v1.ServiceRequestContext} returns this
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.setOrganization =
  function (value) {
    return jspb.Message.setWrapperField(this, 3, value);
  };

/**
 * Clears the message field making it undefined.
 * @return {!proto.udemy.dto.service_request_context.v1.ServiceRequestContext} returns this
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.clearOrganization =
  function () {
    return this.setOrganization(undefined);
  };

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.hasOrganization =
  function () {
    return jspb.Message.getField(this, 3) != null;
  };

/**
 * optional RequestContextAttributions attributions = 4;
 * @return {?proto.udemy.dto.service_request_context.v1.RequestContextAttributions}
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.getAttributions =
  function () {
    return /** @type{?proto.udemy.dto.service_request_context.v1.RequestContextAttributions} */ (
      jspb.Message.getWrapperField(
        this,
        udemy_dto_service_request_context_v1beta_request_context_attributions_pb.RequestContextAttributions,
        4
      )
    );
  };

/**
 * @param {?proto.udemy.dto.service_request_context.v1.RequestContextAttributions|undefined} value
 * @return {!proto.udemy.dto.service_request_context.v1.ServiceRequestContext} returns this
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.setAttributions =
  function (value) {
    return jspb.Message.setWrapperField(this, 4, value);
  };

/**
 * Clears the message field making it undefined.
 * @return {!proto.udemy.dto.service_request_context.v1.ServiceRequestContext} returns this
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.clearAttributions =
  function () {
    return this.setAttributions(undefined);
  };

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.udemy.dto.service_request_context.v1.ServiceRequestContext.prototype.hasAttributions =
  function () {
    return jspb.Message.getField(this, 4) != null;
  };

goog.object.extend(exports, proto.udemy.dto.service_request_context.v1);
