// @udemy/protobuf-objects/build/protos/udemy/dto/organization/v1beta/organization_pb.js
// source: udemy/dto/organization/v1beta/organization.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require("google-protobuf");
var goog = jspb;
var global = Function("return this")();

var google_protobuf_timestamp_pb = require("google-protobuf/google/protobuf/timestamp_pb.js");
goog.object.extend(proto, google_protobuf_timestamp_pb);
var udemy_dto_organization_v1beta_organization_features_pb = require("../../../../udemy/dto/organization/v1beta/organization_features_pb.js");
goog.object.extend(
  proto,
  udemy_dto_organization_v1beta_organization_features_pb
);
var udemy_dto_organization_v1beta_organization_type_pb = require("../../../../udemy/dto/organization/v1beta/organization_type_pb.js");
goog.object.extend(proto, udemy_dto_organization_v1beta_organization_type_pb);
goog.exportSymbol("proto.udemy.dto.organization.v1.Organization", null, global);
goog.exportSymbol(
  "proto.udemy.dto.organization.v1.UserOrganizationRole",
  null,
  global
);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.udemy.dto.organization.v1.Organization = function (opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(proto.udemy.dto.organization.v1.Organization, jspb.Message);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.udemy.dto.organization.v1.Organization.displayName =
    "proto.udemy.dto.organization.v1.Organization";
}

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * Optional fields that are not set will be set to undefined.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
   * @param {boolean=} opt_includeInstance Deprecated. whether to include the
   *     JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.udemy.dto.organization.v1.Organization.prototype.toObject = function (
    opt_includeInstance
  ) {
    return proto.udemy.dto.organization.v1.Organization.toObject(
      opt_includeInstance,
      this
    );
  };

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Deprecated. Whether to include
   *     the JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.udemy.dto.organization.v1.Organization} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.udemy.dto.organization.v1.Organization.toObject = function (
    includeInstance,
    msg
  ) {
    var f,
      obj = {
        instant:
          (f = msg.getInstant()) &&
          google_protobuf_timestamp_pb.Timestamp.toObject(includeInstance, f),
        organizationId: jspb.Message.getFieldWithDefault(msg, 2, 0),
        identifier: jspb.Message.getFieldWithDefault(msg, 3, ""),
        domain: jspb.Message.getFieldWithDefault(msg, 4, ""),
        features:
          (f = msg.getFeatures()) &&
          udemy_dto_organization_v1beta_organization_features_pb.OrganizationFeatures.toObject(
            includeInstance,
            f
          ),
        type: jspb.Message.getFieldWithDefault(msg, 6, 0),
      };

    if (includeInstance) {
      obj.$jspbMessageInstance = msg;
    }
    return obj;
  };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.udemy.dto.organization.v1.Organization}
 */
proto.udemy.dto.organization.v1.Organization.deserializeBinary = function (
  bytes
) {
  var reader = new jspb.BinaryReader(bytes);
  var msg = new proto.udemy.dto.organization.v1.Organization();
  return proto.udemy.dto.organization.v1.Organization.deserializeBinaryFromReader(
    msg,
    reader
  );
};

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.udemy.dto.organization.v1.Organization} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.udemy.dto.organization.v1.Organization}
 */
proto.udemy.dto.organization.v1.Organization.deserializeBinaryFromReader =
  function (msg, reader) {
    while (reader.nextField()) {
      if (reader.isEndGroup()) {
        break;
      }
      var field = reader.getFieldNumber();
      switch (field) {
        case 1:
          var value = new google_protobuf_timestamp_pb.Timestamp();
          reader.readMessage(
            value,
            google_protobuf_timestamp_pb.Timestamp.deserializeBinaryFromReader
          );
          msg.setInstant(value);
          break;
        case 2:
          var value = /** @type {number} */ (reader.readUint64());
          msg.setOrganizationId(value);
          break;
        case 3:
          var value = /** @type {string} */ (reader.readString());
          msg.setIdentifier(value);
          break;
        case 4:
          var value = /** @type {string} */ (reader.readString());
          msg.setDomain(value);
          break;
        case 5:
          var value =
            new udemy_dto_organization_v1beta_organization_features_pb.OrganizationFeatures();
          reader.readMessage(
            value,
            udemy_dto_organization_v1beta_organization_features_pb
              .OrganizationFeatures.deserializeBinaryFromReader
          );
          msg.setFeatures(value);
          break;
        case 6:
          var value =
            /** @type {!proto.udemy.dto.organization.v1.OrganizationType} */ (
              reader.readEnum()
            );
          msg.setType(value);
          break;
        default:
          reader.skipField();
          break;
      }
    }
    return msg;
  };

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.udemy.dto.organization.v1.Organization.prototype.serializeBinary =
  function () {
    var writer = new jspb.BinaryWriter();
    proto.udemy.dto.organization.v1.Organization.serializeBinaryToWriter(
      this,
      writer
    );
    return writer.getResultBuffer();
  };

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.udemy.dto.organization.v1.Organization} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.udemy.dto.organization.v1.Organization.serializeBinaryToWriter =
  function (message, writer) {
    var f = undefined;
    f = message.getInstant();
    if (f != null) {
      writer.writeMessage(
        1,
        f,
        google_protobuf_timestamp_pb.Timestamp.serializeBinaryToWriter
      );
    }
    f = message.getOrganizationId();
    if (f !== 0) {
      writer.writeUint64(2, f);
    }
    f = message.getIdentifier();
    if (f.length > 0) {
      writer.writeString(3, f);
    }
    f = message.getDomain();
    if (f.length > 0) {
      writer.writeString(4, f);
    }
    f = message.getFeatures();
    if (f != null) {
      writer.writeMessage(
        5,
        f,
        udemy_dto_organization_v1beta_organization_features_pb
          .OrganizationFeatures.serializeBinaryToWriter
      );
    }
    f = message.getType();
    if (f !== 0.0) {
      writer.writeEnum(6, f);
    }
  };

/**
 * optional google.protobuf.Timestamp instant = 1;
 * @return {?proto.google.protobuf.Timestamp}
 */
proto.udemy.dto.organization.v1.Organization.prototype.getInstant =
  function () {
    return /** @type{?proto.google.protobuf.Timestamp} */ (
      jspb.Message.getWrapperField(
        this,
        google_protobuf_timestamp_pb.Timestamp,
        1
      )
    );
  };

/**
 * @param {?proto.google.protobuf.Timestamp|undefined} value
 * @return {!proto.udemy.dto.organization.v1.Organization} returns this
 */
proto.udemy.dto.organization.v1.Organization.prototype.setInstant = function (
  value
) {
  return jspb.Message.setWrapperField(this, 1, value);
};

/**
 * Clears the message field making it undefined.
 * @return {!proto.udemy.dto.organization.v1.Organization} returns this
 */
proto.udemy.dto.organization.v1.Organization.prototype.clearInstant =
  function () {
    return this.setInstant(undefined);
  };

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.udemy.dto.organization.v1.Organization.prototype.hasInstant =
  function () {
    return jspb.Message.getField(this, 1) != null;
  };

/**
 * optional uint64 organization_id = 2;
 * @return {number}
 */
proto.udemy.dto.organization.v1.Organization.prototype.getOrganizationId =
  function () {
    return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 2, 0));
  };

/**
 * @param {number} value
 * @return {!proto.udemy.dto.organization.v1.Organization} returns this
 */
proto.udemy.dto.organization.v1.Organization.prototype.setOrganizationId =
  function (value) {
    return jspb.Message.setProto3IntField(this, 2, value);
  };

/**
 * optional string identifier = 3;
 * @return {string}
 */
proto.udemy.dto.organization.v1.Organization.prototype.getIdentifier =
  function () {
    return /** @type {string} */ (
      jspb.Message.getFieldWithDefault(this, 3, "")
    );
  };

/**
 * @param {string} value
 * @return {!proto.udemy.dto.organization.v1.Organization} returns this
 */
proto.udemy.dto.organization.v1.Organization.prototype.setIdentifier =
  function (value) {
    return jspb.Message.setProto3StringField(this, 3, value);
  };

/**
 * optional string domain = 4;
 * @return {string}
 */
proto.udemy.dto.organization.v1.Organization.prototype.getDomain = function () {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 4, ""));
};

/**
 * @param {string} value
 * @return {!proto.udemy.dto.organization.v1.Organization} returns this
 */
proto.udemy.dto.organization.v1.Organization.prototype.setDomain = function (
  value
) {
  return jspb.Message.setProto3StringField(this, 4, value);
};

/**
 * optional OrganizationFeatures features = 5;
 * @return {?proto.udemy.dto.organization.v1.OrganizationFeatures}
 */
proto.udemy.dto.organization.v1.Organization.prototype.getFeatures =
  function () {
    return /** @type{?proto.udemy.dto.organization.v1.OrganizationFeatures} */ (
      jspb.Message.getWrapperField(
        this,
        udemy_dto_organization_v1beta_organization_features_pb.OrganizationFeatures,
        5
      )
    );
  };

/**
 * @param {?proto.udemy.dto.organization.v1.OrganizationFeatures|undefined} value
 * @return {!proto.udemy.dto.organization.v1.Organization} returns this
 */
proto.udemy.dto.organization.v1.Organization.prototype.setFeatures = function (
  value
) {
  return jspb.Message.setWrapperField(this, 5, value);
};

/**
 * Clears the message field making it undefined.
 * @return {!proto.udemy.dto.organization.v1.Organization} returns this
 */
proto.udemy.dto.organization.v1.Organization.prototype.clearFeatures =
  function () {
    return this.setFeatures(undefined);
  };

/**
 * Returns whether this field is set.
 * @return {boolean}
 */
proto.udemy.dto.organization.v1.Organization.prototype.hasFeatures =
  function () {
    return jspb.Message.getField(this, 5) != null;
  };

/**
 * optional OrganizationType type = 6;
 * @return {!proto.udemy.dto.organization.v1.OrganizationType}
 */
proto.udemy.dto.organization.v1.Organization.prototype.getType = function () {
  return /** @type {!proto.udemy.dto.organization.v1.OrganizationType} */ (
    jspb.Message.getFieldWithDefault(this, 6, 0)
  );
};

/**
 * @param {!proto.udemy.dto.organization.v1.OrganizationType} value
 * @return {!proto.udemy.dto.organization.v1.Organization} returns this
 */
proto.udemy.dto.organization.v1.Organization.prototype.setType = function (
  value
) {
  return jspb.Message.setProto3EnumField(this, 6, value);
};

/**
 * @enum {number}
 */
proto.udemy.dto.organization.v1.UserOrganizationRole = {
  USER_ORGANIZATION_ROLE_UNSPECIFIED: 0,
  USER_ORGANIZATION_ROLE_ADMIN: 1,
  USER_ORGANIZATION_ROLE_OWNER: 2,
  USER_ORGANIZATION_ROLE_GROUP_ADMIN: 3,
  USER_ORGANIZATION_ROLE_STUDENT: 4,
};

goog.object.extend(exports, proto.udemy.dto.organization.v1);
