// @udemy/protobuf-objects/build/protos/udemy/dto/organization/v1beta/organization_features_pb.js
// source: udemy/dto/organization/v1beta/organization_features.proto
/**
 * @fileoverview
 * @enhanceable
 * @suppress {missingRequire} reports error on implicit type usages.
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!
/* eslint-disable */
// @ts-nocheck

var jspb = require("google-protobuf");
var goog = jspb;
var global = Function("return this")();

goog.exportSymbol(
  "proto.udemy.dto.organization.v1.OrganizationFeatures",
  null,
  global
);
/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.udemy.dto.organization.v1.OrganizationFeatures = function (opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null);
};
goog.inherits(
  proto.udemy.dto.organization.v1.OrganizationFeatures,
  jspb.Message
);
if (goog.DEBUG && !COMPILED) {
  /**
   * @public
   * @override
   */
  proto.udemy.dto.organization.v1.OrganizationFeatures.displayName =
    "proto.udemy.dto.organization.v1.OrganizationFeatures";
}

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * Optional fields that are not set will be set to undefined.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     net/proto2/compiler/js/internal/generator.cc#kKeyword.
   * @param {boolean=} opt_includeInstance Deprecated. whether to include the
   *     JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.udemy.dto.organization.v1.OrganizationFeatures.prototype.toObject =
    function (opt_includeInstance) {
      return proto.udemy.dto.organization.v1.OrganizationFeatures.toObject(
        opt_includeInstance,
        this
      );
    };

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Deprecated. Whether to include
   *     the JSPB instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.udemy.dto.organization.v1.OrganizationFeatures} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.udemy.dto.organization.v1.OrganizationFeatures.toObject = function (
    includeInstance,
    msg
  ) {
    var f,
      obj = {
        learningPathProPath: jspb.Message.getBooleanFieldWithDefault(
          msg,
          1,
          false
        ),
        isUbProLicenseAssignmentEnabled:
          jspb.Message.getBooleanFieldWithDefault(msg, 2, false),
        hideOrgWideLearnerData: jspb.Message.getBooleanFieldWithDefault(
          msg,
          3,
          false
        ),
      };

    if (includeInstance) {
      obj.$jspbMessageInstance = msg;
    }
    return obj;
  };
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.udemy.dto.organization.v1.OrganizationFeatures}
 */
proto.udemy.dto.organization.v1.OrganizationFeatures.deserializeBinary =
  function (bytes) {
    var reader = new jspb.BinaryReader(bytes);
    var msg = new proto.udemy.dto.organization.v1.OrganizationFeatures();
    return proto.udemy.dto.organization.v1.OrganizationFeatures.deserializeBinaryFromReader(
      msg,
      reader
    );
  };

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.udemy.dto.organization.v1.OrganizationFeatures} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.udemy.dto.organization.v1.OrganizationFeatures}
 */
proto.udemy.dto.organization.v1.OrganizationFeatures.deserializeBinaryFromReader =
  function (msg, reader) {
    while (reader.nextField()) {
      if (reader.isEndGroup()) {
        break;
      }
      var field = reader.getFieldNumber();
      switch (field) {
        case 1:
          var value = /** @type {boolean} */ (reader.readBool());
          msg.setLearningPathProPath(value);
          break;
        case 2:
          var value = /** @type {boolean} */ (reader.readBool());
          msg.setIsUbProLicenseAssignmentEnabled(value);
          break;
        case 3:
          var value = /** @type {boolean} */ (reader.readBool());
          msg.setHideOrgWideLearnerData(value);
          break;
        default:
          reader.skipField();
          break;
      }
    }
    return msg;
  };

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.udemy.dto.organization.v1.OrganizationFeatures.prototype.serializeBinary =
  function () {
    var writer = new jspb.BinaryWriter();
    proto.udemy.dto.organization.v1.OrganizationFeatures.serializeBinaryToWriter(
      this,
      writer
    );
    return writer.getResultBuffer();
  };

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.udemy.dto.organization.v1.OrganizationFeatures} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.udemy.dto.organization.v1.OrganizationFeatures.serializeBinaryToWriter =
  function (message, writer) {
    var f = undefined;
    f = message.getLearningPathProPath();
    if (f) {
      writer.writeBool(1, f);
    }
    f = message.getIsUbProLicenseAssignmentEnabled();
    if (f) {
      writer.writeBool(2, f);
    }
    f = message.getHideOrgWideLearnerData();
    if (f) {
      writer.writeBool(3, f);
    }
  };

/**
 * optional bool learning_path_pro_path = 1;
 * @return {boolean}
 */
proto.udemy.dto.organization.v1.OrganizationFeatures.prototype.getLearningPathProPath =
  function () {
    return /** @type {boolean} */ (
      jspb.Message.getBooleanFieldWithDefault(this, 1, false)
    );
  };

/**
 * @param {boolean} value
 * @return {!proto.udemy.dto.organization.v1.OrganizationFeatures} returns this
 */
proto.udemy.dto.organization.v1.OrganizationFeatures.prototype.setLearningPathProPath =
  function (value) {
    return jspb.Message.setProto3BooleanField(this, 1, value);
  };

/**
 * optional bool is_ub_pro_license_assignment_enabled = 2;
 * @return {boolean}
 */
proto.udemy.dto.organization.v1.OrganizationFeatures.prototype.getIsUbProLicenseAssignmentEnabled =
  function () {
    return /** @type {boolean} */ (
      jspb.Message.getBooleanFieldWithDefault(this, 2, false)
    );
  };

/**
 * @param {boolean} value
 * @return {!proto.udemy.dto.organization.v1.OrganizationFeatures} returns this
 */
proto.udemy.dto.organization.v1.OrganizationFeatures.prototype.setIsUbProLicenseAssignmentEnabled =
  function (value) {
    return jspb.Message.setProto3BooleanField(this, 2, value);
  };

/**
 * optional bool hide_org_wide_learner_data = 3;
 * @return {boolean}
 */
proto.udemy.dto.organization.v1.OrganizationFeatures.prototype.getHideOrgWideLearnerData =
  function () {
    return /** @type {boolean} */ (
      jspb.Message.getBooleanFieldWithDefault(this, 3, false)
    );
  };

/**
 * @param {boolean} value
 * @return {!proto.udemy.dto.organization.v1.OrganizationFeatures} returns this
 */
proto.udemy.dto.organization.v1.OrganizationFeatures.prototype.setHideOrgWideLearnerData =
  function (value) {
    return jspb.Message.setProto3BooleanField(this, 3, value);
  };

goog.object.extend(exports, proto.udemy.dto.organization.v1);
