// @udemy/react-header/dist/esm/ufb-desktop/logged-out-header.react-component.js
import { observer } from "mobx-react";
import React from "react";
import { AboveTheFoldProvider } from "@udemy/design-system-utils";
import { useI18n } from "@udemy/i18n";
import { UFBLogo } from "@udemy/react-brand-components";
import { Button } from "@udemy/react-core-components";
import { loadCommonAppContext } from "../external/load-common-app-context";
import { HeaderStore } from "../header.mobx-store";
import styles from "./logged-out-header.module.css";
export const LoggedOutHeader = observer((props) => {
  const { gettext, locale } = useI18n();
  const [headerStore] = React.useState(() => new HeaderStore(props));
  React.useEffect(() => {
    loadCommonAppContext(locale, true).then((response) => {
      headerStore.setUserSpecificContext(response.data.header);
    });
  }, [headerStore, locale]);
  const { organizationState } = headerStore.userSpecificContext;
  return /*#__PURE__*/ React.createElement(
    AboveTheFoldProvider,
    null,
    /*#__PURE__*/ React.createElement(
      "div",
      {
        className: styles["header"],
      },
      /*#__PURE__*/ React.createElement(
        "a",
        {
          href: headerStore.urls.BROWSE,
        },
        /*#__PURE__*/ React.createElement(UFBLogo, null)
      ),
      /*#__PURE__*/ React.createElement("div", {
        style: {
          flex: 1,
        },
      }),
      organizationState &&
        /*#__PURE__*/ React.createElement(
          Button,
          {
            size: "medium",
            udStyle: "primary",
            componentClass: "a",
            href: `/?next=${encodeURIComponent(window.location.href)}`,
          },
          organizationState.is_forced_sso
            ? gettext("Log in")
            : gettext("Log in or sign up")
        )
    )
  );
});
