// @udemy/shopping/dist/esm/wishlist/wishlist.react-component.js
var _class;
import { observer } from "mobx-react";
import React from "react";
import { withI18n } from "@udemy/i18n";
import { withUDData } from "@udemy/ud-data";
import { WishlistButton } from "./wishlist-button.react-component";
let InternalWishlist =
  observer(
    (_class = class InternalWishlist extends React.Component {
      render() {
        const {
          wishlistStore,
          round,
          isMobile,
          // isCourseInUserSubscription,
          // uiRegion,
          udData,
          gettext,
          ...givenButtonProps
        } = this.props;
        const { Config, me } = udData;

        // Commenting out this for now since we don't need this for LOHP
        // Todo: Refactor to pull this out of Wishlist Component; instead, toggle at the higher
        // level between wishlist and SaveToListButton
        // if (this.isPersonalPlanSubscriber && isCourseInUserSubscription) {
        //     return (
        //         <SaveToListButtonWrapper
        //             course={{id: wishlistStore.course.id, is_in_user_subscription: true}}
        //             uiRegion={uiRegion}
        //             size="medium"
        //             labelPosition="left"
        //         />
        //     );
        // }

        if (!Config.features.wishlist || me.isLoading) {
          return null;
        }
        return /*#__PURE__*/ React.createElement(
          "div",
          null,
          /*#__PURE__*/ React.createElement(
            WishlistButton,
            Object.assign(
              {
                "data-purpose": "toggle-wishlist",
                isWishlisted: wishlistStore.isWishlisted,
                isLoading: wishlistStore.apiRequestIsLoading,
                onClick: wishlistStore.toggleWishlist,
                round: round,
                wishlistCta: isMobile
                  ? gettext("Add to Wishlist")
                  : gettext("Wishlist"),
              },
              givenButtonProps
            )
          )
        );
      }
    })
  ) || _class;
export const Wishlist = withI18n(withUDData(InternalWishlist));
